KEP-4317: Pod Certificates (#4318)
kubernetes/enhancements: keps/sig-auth/4317-pod-certificates/README.md

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary and Motivation

The certificates.k8s.io API group provides a flexible, pluggable mechanism to request X.509 certificates within a Kubernetes cluster, but actually making certificates available to your workloads is left as an exercise for the reader.

This KEP defines three pieces of fundamental machinery:
* An X.509 extension for expressing pod identity
* Node restriction support for certificate requests that use the PodIdentity
  extension.
* A PodCertificate projected volume source that instructs the kubelet to handle
  provisioning a key and certificate chain on behalf of a pod.  
Taken together, this machinery makes it feasible to securely and automatically
deliver X.509 certificates to every pod in a cluster, without imposing an
unreasonable burden on application developers or cluster administrators.  

> As a first application of this machinery, this KEP defines additional machinery that allows pods to automatically receive mTLS client certificates suitable for authenticating to kube-apiserver, iwth the same attributes provided by bound service account tokens:

Reviewer: ```suggestion
As a first application of this machinery, this KEP defines additional machinery that allows pods to automatically receive mTLS client certificates suitable for authenticating to kube-apiserver, with the same attributes provided by bound service account tokens:
```

Author: Done.

* A new TLS user-mapping capability in kube-apiserver to understand client
  certificates that embed PodIdentity extensions.
* A new approver/signer pair, shipped in kube-controller-manager, to issue
  compatible client certificate to pods.
* An enhancement to client-go's InClusterConfig that automatically loads a
  credential bundle from a well-known filesystem path and uses is as a client
  certificate.

There is a draft implementation of the design outlined in this KEP:
https://github.com/kubernetes/kubernetes/pull/121596

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

> ### Goals

Author: Additional goal: revocation for the kube-apiserver pod client certificate case --- revocation handled via pod binding.

Author: Added a section for this.

> **Ensure that certificate issuance is accomplished without relying on bearer tokens at any step**:  One of the most significant benefits of mTLS authentication is that the credential is split into two pieces (private key and certificate), only one of which is ever transmitted to the counterparty (and thus subject to eavesdropping or theft by the counterparty).  In order to preserve this guarantee, it's important that bearer tokens are not involved *at any step*, otherwise the security of the scheme devolves to just being a bearer token.  In this KEP, as long as both kubelet and the signer implementation authenticate to kube-apiserver using their own separately-bootstrapped client certificates, then no bearer tokens are involved in certificate issuance.

Reviewer 2: The API server need not be the guarantor here. It could be the cluster administrator looking at their authn / authz design and concluding that bearer tokens aren't involved in this path.

Maybe also “can't be involved“; that'd be an even stronger promise or guarantee.

> **Ensure that certificate issuance is accomplished without relying on bearer tokens at any step**:  One of the most significant benefits of mTLS authentication is that the credential is split into two pieces (private key and certificate), only one of which is ever transmitted to the counterparty (and thus subject to eavesdropping or theft by the counterparty).  In order to preserve this guarantee, it's important that bearer tokens are not involved *at any step*, otherwise the security of the scheme devolves to just being a bearer token.  In this KEP, as long as both kubelet and the signer implementation authenticate to kube-apiserver using their own separately-bootstrapped client certificates, then no bearer tokens are involved in certificate issuance.

Author: Yeah, I guess my intention here was to say that this should be possible as long as the admin sets everything up correctly.  I will clarify.

> **Ensure that certificate issuance is accomplished without relying on bearer tokens at any step**:  One of the most significant benefits of mTLS authentication is that the credential is split into two pieces (private key and certificate), only one of which is ever transmitted to the counterparty (and thus subject to eavesdropping or theft by the counterparty).  In order to preserve this guarantee, it's important that bearer tokens are not involved *at any step*, otherwise the security of the scheme devolves to just being a bearer token.  In this KEP, as long as both kubelet and the signer implementation authenticate to kube-apiserver using their own separately-bootstrapped client certificates, then no bearer tokens are involved in certificate issuance.

Author: Updated.

**Make it easy to ship secure third-party signer implementations**:  Securely implementing and operating a third-party signer implementation today requires intimate familiarity with Kubernetes' security model, otherwise your signer may negate Kubernetes' node restriction security boundary.  This KEP moves responsibilty for enforcing node restriction on CSRs into kube-apiserver, so that signer implementations do not need to consider it.

### Non-Goals

**Specify a solution for pod-to-pod mTLS in core Kubernetes**: Pod-to-pod mTLS (aka service mesh) is a rapidly-evolving area.  While the PodCertificate volume and PodIdentity X.509 extension are expected (and designed to be) useful for service mesh implementations, it doesn't make sense to try and standardize pod-to-pod mTLS at the core Kubernetes layer today.

> **Support TPM-backed private keys**: In this proposal, private keys are

Author: This KEP should have a simple design of how this could eventually work, to show that we are not closing off any important avenues for TPM support.

Author: Section added, but not yet filled out.

contained to the node (held only in kubelet memory and tmpfs projected volumes),
but they can still be extracted from a compromised workload (or compromised

> node).  Storing the private keys in TPMs closes this loophole, but will require

Author: I wasn't being precise here, sorry.  My intention was just to vaguely handwave at some sort of HSM support.

However, I have been thinking a little bit more about this as I'm preparing to ramp back up on this after working on external token signing.  I am starting to think that it might be a good idea to *not* write the key into the pod filesystem, and instead expose some external signature interface via UDS in the pod.   Basically, start by requiring external TLS.

To establish a TLS connection to kube-apiserver, we would ship a custom crypto.Signer implementation with client-go that knows how to speak to kubelet via the UDS and request signatures of data using the per-pod key held by kubelet.

In Alpha, kubelet wouldn't have to *do* anything fancy with the keys, but starting out with the key not exposed to the pod will make an eventual TPM / HSM transition transparent, and it will also significantly improve the baseline security of the feature.

> node).  Storing the private keys in TPMs closes this loophole, but will require

Author: After API review, we decided to defer any HSM / external signing  support to a future iteration.

an additional KEP to define how kubelet can use the Linux (or other platform's)
virtual TPM subsystem to provide virtual TPMs to pods

**Be 100% compatible with software in the field today**: The primary example in this KEP is that the private key and certificate chain are delivered in a single credential-bundle file.  This is based on implementation and operation experience of a CSI-based certificates solution that delivered private keys and certificate chains in separate files, as expected by most commodity webservers today.  While this is an attractive idea, automatically rotating these separate files is fraught.  Even if both file's content is updated atomically, skewed reads in the workload can result in hard-to-diagnose errors if the workload reads the key before rotation and the certificate chain after rotation (or vice-versa).

**Support private keys shared across multiple pods**: Sharing private keys and certificates across multiple pods may be required in certain applications.  However, this use case is well-served by Secret projected volume sources.

> **Support human-in-the-loop approval for certificate requests**: This KEP assumes that CSR approval and issuance is fully-automated, because this process blocks pod startup for pods that use PodCertificate projected volume sourcs.

Author: Done.

> 

Reviewer 3: Probable non-goals to add, that have cropped up as concerns in parallel impls (e.g. SPIRE) that are by-definition out of scope for this as written:

* Supporting identity delegation (minting certs that can be used by workloads for intermediate CA purposes)
* Supporting identity impersonation (one privileged workload minting many certs on-behalf-of other workloads)

Author: Captured as an unresolved block in the non-goals section.

**Standardize on using the SPIFFE SVID format in core Kubernetes**: The certificates issued by `kubernetes.io/kube-apiserver-client-pod` can be made compatible with SPIFFE in the future, since they currently don't include any Subject Alternate Name entries.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

#### Story 1

> I'm an application developer building a Kubernetes controller that I want to deploy into my cluster.  I want to make sure that my controller securely authenticates to kube-apiserver using mTLS.

Author: For the kube-apiserver client certificate case, the intention is to handle revocation via a combination of the pod binding concept and short certificate lifetime.  Delete the pod, and all of its client certificates are no longer trusted.

I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `noAutomountServiceAccountToken` because I want to ensure that my pod cannot use bearer tokens for authentication.
```yaml
apiVersion: v1
kind: Pod
metadata:
  namespace: default
  name: pod-certificates-example
spec:
  restartPolicy: OnFailure
  automountServiceAccountToken: false
  containers:
  - name: main
    image: debian
    command: ['sleep', 'infinity']
    volumeMounts:
    - name: kube-apiserver-client-certificate
      mountPath: /var/run/secrets/kubernetes.io/api-mtls-access
  volumes:
  - name: kube-apiserver-client-certificate
    projected:
      sources:
      - podCertificate:

>           signerName: "kubernetes.io/kube-apiserver-client-pod"

Reviewer 4: > CN is username, Organizations are groups

This is where I thought we could use custom Subject RDN key names that I talked about during the api-review mtg. It might be a bit more flexible than an ASN.1 extension.

>           signerName: "kubernetes.io/kube-apiserver-client-pod"

Author: OK, I'm about to upload a revision that adds a new PodCertificateRequest type.  I've left a bunch of UNRESOLVED blocks to update details of the signer.

>           signerName: "kubernetes.io/kube-apiserver-client-pod"

Author: > This is where I thought we could use custom Subject RDN key names that I talked about during the api-review mtg - if such a need arose. It might be a bit more flexible than an ASN.1 extension.

I've done some googling, and I don't really understand what the usability difference in practice would be.  Both seem to require a custom OID, and non-standard parsing logic for the certificate.

          credentialBundlePath: client-credentials.pem
      - configMap:
          localObjectReference: kube-root-ca.crt
          items:
          - key: ca.crt
            path: kube-apiserver-root-certificate.pem
      - downwardAPI:
          items:
          - path: namespace
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
```

I ensure that my pod uses client-go's InClusterConfig, and client-go automatically uses the mTLS client certificate to authenticate to kube-apiserver.

### Risks and Mitigations

**Scalability**: In the current design and draft implementation, kubelet holds
the private keys for each pod on the node in memory.  If kubelet restarts, then
the private keys are lost, and kubelet will need to immediately generate new
keys and request new certificates for each volume on the pod.  If kubelet
crashloops, this can generate high CertificateSigningRequest traffic on
kube-apiserver.  This can potentially be mitigated by giving kubelet the

> capability to read back the existing key from the projected volume on startup.

Author: I'm not sure.  I don't *think* that that would be a problem --- if an attacker somehow stole a private key from pod A, then they could use this trick to get a new certificate for that private key from another compromised pod B.  How big of a deal is that?

I'm inclined to say that a Kubelet restart requires a re-key for now.

> capability to read back the existing key from the projected volume on startup.

Author: I'm actually not sure how to make files from this specific projected volume source readonly --- the whole projected volume mount could be made readonly, but that would affect other files from other sources in the same volume.

> capability to read back the existing key from the projected volume on startup.

Reviewer 3: > if an attacker somehow stole a private key from pod A, then they could use this trick to get a new certificate for that private key from another compromised pod B. How big of a deal is that?

It allows for impersonating pod A (steal privkey from your volume) from pod B (write privkey to my volume, kubelet), which isn't great. 

Kubelet should probably simply never trust anything once it's been mounted into a pod, full stop, even if that creates churn on restart - that's sort of the price of putting the `kubelet` in the middle.

## Design Details

### X.509 PodIdentity Extension

The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
identity information into an X.509 certificate or certificate request, we can
define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
define a management policy for the PEN) and specify that it contains the
following DER-encoded structure:

> ```go

Author: I'll do my best to figure out the ASN.1 notation for it.

"Registration" is just us assigning an OID under the Kubernetes Private Enterprise Number via this KEP.  I don't see why we couldn't make backwards-compatible modifications, similar to any other Kubernetes API.  But we can also always allocate a new OID.

type podIdentityASN1 struct {
	Namespace          string `asn1:"utf8"`

> 	ServiceAccountName string `asn1:"utf8"`

Author: I've put an unresolved block in the API object section discussing this.  The root of the issue is that Kubelet doesn't know the Service Account or Node UIDs, and so can't put them in the PodCertificateRequest object.

> 	ServiceAccountName string `asn1:"utf8"`

Author: I figured out how to do this properly.  It's now specified.

	PodName            string `asn1:"utf8"`
	PodUID             string `asn1:"utf8"`
	NodeName           string `asn1:"utf8"`
}
```

> Certificates and certificate requests that embed this extension do not need to include any data in the standard X.509 name or subjectAlternateName fields. The CertificateSigningRequests generated by the kubelet will not, nor will the certificates issued by `kubernetes.io/kube-apiserver-client-pod`.

Author: After API review, we decided that the certificates issued by `kubernetes.io/kube-apiserver-client-pod`  should also have a backwards-compatible CommonName (like `system:serviceaccount:foo:bar`), so at least in those certs this should not be a critical extension.

(The extension as a whole is not critical/non-critical --- it's only when you embed it into a certificate that you declare whether it is a critical or non-critical extension).

Utility functions for embedding and extracting this extension from `crypto/x509`
Certificate and CertificateRequest objects are defined in
`k8s.io/component-helpers/kubernetesx509`.  They are intended to be used both by
core Kubernetes components and third-party software that needs to interact with
these certificates.

### PodIdentity Extension Node Restriction

> The noderestriction admission plugin will check the consistency of PodIdentity extensions in CertificateSigningRequest objects created by `system:node:` identities.  The creation request will only be admitted if the PodIdentity extension refers to a pod that is currently running on the node that requested the certificate.

Author: The node restriction plugin is built in to kube-apiserver.  It can probably be disabled, but I don't think it is on most distributions?

> The noderestriction admission plugin will check the consistency of PodIdentity extensions in CertificateSigningRequest objects created by `system:node:` identities.  The creation request will only be admitted if the PodIdentity extension refers to a pod that is currently running on the node that requested the certificate.

Reviewer 3: Okay, so the `noderestriction` admission plugin is the default "floor" and people can always add more strict/granular admission plugins on top. IIUC, that makes sense.

actually been scheduled onto that node.

<<[UNRESOLVED]>>
Is it necessary or desirable to forbid non-kubelet identities from creating CertificateSigningRequest objects that embed a PodIdentity extension?
<<[/UNRESOLVED]>>

During Alpha, this capability will be behind the PodIdentityX509Support feature gate.

### PodCertificate Projected Volume Sources

The PodCertificate projected volume source instructs kubelet to inject a
credential bundle consisting of a private key and certificate chain into the
container fileystem.  kubelet handles generating the key, requesting a
certificate from the named signer, and renewing the certificate before it
expires.

The kubelet supports only a few key generation strategies, each named by a string
that encapsulates both the key type and any parameters necessary (for example,
RSA modulus size).  The intention is to offer a a tasting menu of reasonable key
choices, rather than offering a flexibility to a wide variety of parameters.
Key types supported by kubelet are "RSA2048", "RSA3072", "RSA4096", "ECDSAP256",
and "ECDSAP384".  If no key type is specified, kubelet defaults to "ECDSAP256".

> 

Author: The problem here is that Kubelet has to generate the key, so fundamentally if we want to allow the user to exercise option X, then Kubelet needs to implement option X.

However, I'm not sure there really are any other choices to make for private keys.  RSA lets you pick the modulus, which is specified here, and ECDSA lets you pick a curve, which is specified here.

> 

Reviewer 3: > The problem here is that Kubelet has to generate the key, so fundamentally if we want to allow the user to exercise option X, then Kubelet needs to implement option X.

Yes, that's a bit unfortunate. I do understand it's an unavoidable limitation of the goals/non-goals here tho.

<<[UNRESOLVED]>>
Should RSA2048 be offered?  @ahmedtd inclined to say no.  Are there any other key types that should be offered?
<<[/UNRESOLVED]>>

Once the key is generated, kubelet creates a CertificateSigningRequest that
contains an X.509 certificate request with no data in the name or
subjectAlternateName fields, but instead embeds an X.509 PodIdentity extension

> with the details of the pod that mounted the volume.  The signer is intended to

Reviewer 3: See prev. comment about the OID PodIdentity extension - this section either needs some MUSTS or I'm not sure including the PodIdentity extension is much practical use.

- If the PodIdentity extension is a suggestion/hint to the issuer, not a rule - why define it at all as a x509 extension, if there's no guarantee the eventual cert will meaningfully relate to it/carry some/all of it at all?
- If all the fields in PodIdentity are optional (including `PodName`), what is a `named pod`, exactly?

Author: Re Point 1: The intention here is that the signer implementation will receive a fixed-format CSR from Kubelet, extract the pod identity (to the extent that it cares about it) from the CSR, and then issue the certificate it believes the pod should have.  For example, a hypothetical custom service mesh signer implementation would read the PodIdentity extension, but then issue a SPIFFE certificate that encodes the pod's identity in the way that it likes.

This division of responsibilities will keep us from having an explosion of configuration knobs on the certificate volume.

Re Point 2: In this section, it is Kubelet filling out the PodIdentity extension, so it will set all fields.  I'll clarify.

Reviewer 3: >  For example, a hypothetical custom service mesh signer implementation would read the PodIdentity extension, but then issue a SPIFFE certificate that encodes the pod's identity in the way that it likes.

That makes sense. My concern was that if _all_ the fields in the PodIdentity are explicitly optional in the spec, it seems like the spec is encoding a situation where a custom implementation could not get enough information/hints from the provided PodIdentity to even sufficiently identify the pod, which seems not-ideal.

> Re Point 2: In this section, it is Kubelet filling out the PodIdentity extension, so it will set all fields. I'll clarify.

:+1:

Author: After API review, we decided to move to use a new PodCertificateRequest type.  This significantly cuts down the number of parties that need to care about the X.509 extension --- now it's basically only the built-in `kubernetes.io/kube-apiserver-client-pod` signer, and the kube-apiserver client certificate authenticator.

Random third-party signers will now just directly read pod details from top-level fields on the PodCertificateRequest object.

treat this certificate request merely as an indication that it should issue a
certificate for the named pod, rather than trying to issue a certificate that
follows the structure of the certificate request.

The kubelet then starts a single-object informer on the CertificateSigningRequest object it just created, and waits until the CSR is Approved and Issued.  The kubelet then takes the private key it generated, plus the issued certificate chain, and writes them to the filesystem of the relevant Pod volume, concatenated into a single PEM-encoded file.  The first block in the PEM file contains the private key, and the remaining blocks contain the certificate chain in the same order they were in the CSR.

The kubelet strips any inter-block text comments and / or intra-block headers
that were present in the issued certificate chain from the signer.

> The kubelet always specifies a 24-hour TTL on its certificate requests, although of course the signer is free to ignore the requested TTL and return a certificate with any TTL.  The kubelet will use the issued certificate unless the TTL is less than 1 hour from the current time.

Author: This has been obsoleted by shifting to a new resource type, PodCertificateRequest.  (Revision describing this is about to be uploaded).

During the initial volume mount, before the pod has started, any errors returned during the certificate issuance process will block pod startup.

> The kubelet will begin trying to rotate the certificate once less than 50% of its TTL is remaining, or after 24 hours, whichever is sooner.  Errors in the certificate issuance process are treated as non-fatal (logged but not returned from the volume SetUp function) while the original certificate is still valid.  Once the original certificate has expired, errors are will be returned from the volume SetUp function, because it's better for the pod to noisily go unhealthy than to continue to run with an expired certificate.

Author: This has been obsoleted by the shift to a new PodCertificateRequest object that allows the signer to specify when Kubelet should begin refreshing a given certificate.

dRotiCEIGpCQmdMTDYZez6zTfN3E1yuhyw7bSKc2OJj6+MYly6oQwE0T7WVDNqoa
Ev2ZdHZjePkEhXvGRvm1fyxK4h0UJsaT38/COUW/Dal7Fx6VcJeY2QHI28W+NYP3
AOdx/7Hc6bbPcYZI/c1HPolUhoX6huiltGMn3yOLsaGjajwXUoE2aKsy3AHxeIkQ
4UlKQSabpb8CAwEAAaOBqDCBpTAOBgNVHQ8BAf8EBAMCA6gwEwYDVR0lBAwwCgYI
KwYBBQUHAwIwHwYDVR0jBBgwFoAUuQ7U2/jMkK4nY5R2rPvjyCiC2SQwXQYJKwYB
BAGDwlMBBFAwTgwHZGVmYXVsdAwHZGVmYXVsdAwHcGVtLXBvZAwkMjc3YjljNDkt
MzljMy00ZjIzLWIwYjEtZTg4OWVmOTBjMGIyDAtraW5kLXdvcmtlcjANBgkqhkiG
9w0BAQsFAAOCAQEAQd1G2CHdGPv9Lu9/SD7HPNXwI0K4JrXsWxJQ7Q16IVAuuWIY
mOdumO9L5PtTrV1widNm99oMEtztQHZoHHIkkypjfW+BB8XwO5er+FpcUCuvZ3w/
g9Zk9cZUqCdWER0UDl4RCLcS+M6QSq0nUg+JTZKRGJ5fjns3vZHA9Nswp4XRvjt8
yf6RY6oEJoa10QxB7YiNVmVkqLifhkQiF0Bmz7tONbsjceZ030Cs/wjdssnNzRGL
UdFfcDKw4QbMGqntJXPFqu86lk0IJwir+aTk/31yDZX1ZyOK0Y/SI7ZVVDeETTA9
k68gmm9HQCdRI3stW7TC3lB0Cd1XSSMUISIP0g==
-----END CERTIFICATE-----
```

### API Server PodIdentity Extension Support

The kube-apiserver is extended to authenticate TLS client certificates that:

* Are signed by the existing `--client-ca-file` CA,
* Contain empty Name and SubjectAlternateName data, and
* Contain an X.509 PodIdentity extension.

These certificates are translated to an authenticator.Response that has:

> * `User.Name` set to

Author: Pod UID is captured.  I have an unresolved block in the API type section discussing whether or not we can capture service account UID and node UID.

  `system:serviceaccount:{PodIdentity.Namespace}:{PodIdentity.ServiceAccountName}`.
* `User.Extra["authentication.kubernetes.io/pod-name"]` set to
  `{PodIdentity.PodName}`
* `User.Extra["authentication.kubernetes.io/pod-uid"]` set to
  `{PodIdentity.PodUID}`

> * `User.Extra["authentication.kubernetes.io/node-name"]` set to

Reviewer 5: We should also set the node UID here?

Reviewer 5: And also the service account UID? Or I wonder if that'll be used in User.UID instead?

Author: Ditto the conversatin about SA / Node UIDs above.

  `{PodIdentity.NodeName}`

The net effect is that authenticating via a PodIdentity-bearing cert has the
same object-bound behavior as authenticating via a bound service account token.

During Alpha, this feature will be gated by the PodIdentityX509Support feature
gate.

### API Server Pod Client Certificate Approver / Signer

The kube-controller-manager is extended to support a new certficate signer,
`kubernetes.io/kube-apiserver-client-pod`.

The approver for this signer name automatically approves
CertificateSigningRequests that:
* Contain no Name or Subject Alternate Name data, 
* Embed a PodIdentity extension, and
* Are requested by node identities.

(The approver relies on the noderestriction plugin to verify that the pod and
node in the extension are consistent.)

The signer for this signer name issues certificates for all approved CSRs
addressed to the signer name.  The certificates it issues contain no Name or
Subject Alternate Name data, embed a PodIdentity extension, and have key usages
consistent with the certificate being used as a TLS client certificate. It uses

> the same CA key and certificates as `kubernetes.io/kube-apiserver-client`.

Reviewer 2: I'd go further: we **should** use a different signer.

> the same CA key and certificates as `kubernetes.io/kube-apiserver-client`.

Author: Sure, I will update the doc.

> the same CA key and certificates as `kubernetes.io/kube-apiserver-client`.

Author: Wait, I'm confused... the doc already specifies a custom signer (`kubernetes.io/kube-apiserver-client-pod`).  Do you mean we should use a separate CA?

> the same CA key and certificates as `kubernetes.io/kube-apiserver-client`.

Reviewer 4: At the time of commenting (as you can see in the preview), the document suggested signer reuse. But yes, a different signer and a separate key should be used.

- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary and Motivation

The certificates.k8s.io API group provides a flexible, pluggable mechanism to request X.509 certificates within a Kubernetes cluster, but actually making certificates availble to your workloads is left as an exercise for the reader.  This missing

This KEP defines three pieces of fundamental machinery:
* An X.509 extension for expressing pod identity
* Node restriction support for certificate requests that use the PodIdentity
  extension.

> * A PodCertificate projected volume source that instructs Kubelet to handle

Reviewer 2: (nit)
```suggestion
* A PodCertificate projected volume source that instructs the kubelet to handle
```

Reviewer 2: Well, it's (nearly) always lower case. And usually “the kubelet”, but that's common rather than always.

Author: I'll fix these up.

**Make it easy to ship secure third-party signer implementations**:  Securely implementing and operating a third-party signer implementation today requires intimate familiarity with Kubernetes' security model, otherwise your signer may negate Kubernetes' node restriction security boundary.  This KEP moves responsibilty for enforcing node restriction on CSRs into kube-apiserver, so that signer implementations do not need to consider it.

### Non-Goals

> **Specify a solution for pod-to-pod mTLS in core Kubernetes**: Pod-to-pod mTLS (aka service mesh) is a rapidly-evolving area.  While the PodCertificate volume and PodIdentity X.509 extension are expected (and designed to be) useful for service mesh implementations, it doesn't make sense to try and standardize pod-to-pod mTLS at the core Kubernetes layer today.

Reviewer 2: Maybe the non-goal is to _specify_ a particular solution, but we leave room and consider opportunities for this to work with other APIs (such as CertificateSigningRequest, ClusterTrustBundle, ServiceAccount).

How does that sound? Leaving room for someone else to combine these APIs is what I personally would want, and is a soft ask on behalf of [The Scale Factory](https://scalefactory.com/) where I work.

> **Specify a solution for pod-to-pod mTLS in core Kubernetes**: Pod-to-pod mTLS (aka service mesh) is a rapidly-evolving area.  While the PodCertificate volume and PodIdentity X.509 extension are expected (and designed to be) useful for service mesh implementations, it doesn't make sense to try and standardize pod-to-pod mTLS at the core Kubernetes layer today.

Author: Oh, I see.  My goal is definitely to leave room for these mechanisms to work together, and for people to be able to use them for pod-to-pod mTLS.  (I work on a product that will definitely be able to discard a custom CSI driver and a *bunch* of support code once these generic mechanisms are in core Kubernetes)

I just don't think it's prudent to try to force standardization in this area in core Kubernetes (at least in the near-to-medium future).

I will add an explicit goal to clarify this.

I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `noAutomountServiceAccountToken` because I want to ensure that my pod cannot use bearer tokens for authentication.
```yaml
apiVersion: v1
kind: Pod
metadata:
  namespace: default
  name: pod-certificates-example
spec:
  restartPolicy: OnFailure
  automountServiceAccountToken: false
  containers:
  - name: main
    image: debian
    command: ['sleep', 'infinity']
    volumeMounts:
    - name: kube-apiserver-client-certificate
      mountPath: /var/run/secrets/kubernetes.io/api-mtls-access
  volumes:
  - name: kube-apiserver-client-certificate
    projected:
      sources:
      - podCertificate:
          signerName: "kubernetes.io/kube-apiserver-client-pod"
          credentialBundlePath: client-credentials.pem
      - configMap:
          localObjectReference: kube-root-ca.crt
          items:
          - key: ca.crt
            path: kube-apiserver-root-certificate.pem
      - downwardAPI:
          items:
          - path: namespace
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
```

> I ensure that my pod uses client-go's InClusterConfig, and client-go automatically uses the mTLS client certificate to authenticate to kube-apiserver.

Author: My intention was that this KEP would enhance InClusterConfig to use the certificate if mounted at a particular path.  I'll try to make that more clear.

### Risks and Mitigations

**Scalability**: In the current design and draft implementation, Kubelet holds

> the private keys for each pod on the node in memory.  If Kubelet restarts, then

Author: This is just the way the kubelet is structured today.  All data that is written out into a projected volume is held in memory, and the kubelet writes out the data periodically using the atomicwriter utility package.

I think we are probably not going to be able to banish the keys from kubelet memory without going full TPM.  Without a TPM, the kubelet has to hold the key in memory at least temporarily to do the crypto of making the CSR, and during that time period the kubelet's memory might get swapped to disk, etc.

The primary ways we could mitigate the "kubelet restarts -> reissue all certificates" problem are:
1.  Store the keys in a host-level database, probably something based on boltdb.  This *feels* gross, because it seems like it makes the keys easily accessible for any privileged pods or to someone who has root on the node.  However, if you actually think through the analysis, I don't believe it's actually more insecure than allowing the keys to live in Kubelet memory --- root and privileged pods will be able to rummage through Kubelet's memory if they want to.  It is also fairly simple to retrofit into the kubelet's existing projected volume architecture.  This is the alternative we are moving towards in gke-spiffe, which is a CSI driver implementation of this capability.
2. Read the keys back from the pod filesystem.  This is much harder to fit into the existing architecture, and not actually more secure.  This is the alternative we started with in gke-spiffe.

> the private keys for each pod on the node in memory.  If Kubelet restarts, then

Reviewer 2: The kubelet could defer to something like SoftHSM (or, generically, PKCS#11). And then the PKCS#11 implementation either uses hardware or `mlockall()` to protect the key material from being written to the pagefile.

> the private keys for each pod on the node in memory.  If Kubelet restarts, then

Author: After some discussion in SIG Auth, we decided to defer any sort of HSM capability to a later enhancement.

the private keys are lost, and Kubelet will need to immediately generate new
keys and request new certificates for each volume on the pod.  If Kubelet
crashloops, this can generate high CertificateSigningRequest traffic on
kube-apiserver.  This can potentially be mitigated by giving Kubelet the
capability to read back the existing key from the projected volume on startup.

## Design Details

### X.509 PodIdentity Extension

The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
identity information into an X.509 certificate or certificate request, we can
define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
define a management policy for the PEN) and specify that it contains the
following DER-encoded structure:

```go

> type podIdentityASN1 struct {

Reviewer 6: Somewhat off-topic, but... What would it take to network bind these certificates? E.g. assuming we can trust the underlying network to prevent IP spoofing, could we include a node or pod source IP here as well?

Reviewer 3: Can it be checked as an *issuance* gate for the CSR request, as a pod attestation?

Or do we have to take the declared binding at face value, and check it at time-of-use only?

Reviewer 3: Doesn't even have to be `kube-apiserver` that validates during admission, IIUC?

E.g. you could use SPIRE's attesting agent to validate the admission request for the cert and reject if out-of-band validations fail, or any other thing.

	Namespace          string `asn1:"utf8"`
	ServiceAccountName string `asn1:"utf8"`
	PodName            string `asn1:"utf8"`
	PodUID             string `asn1:"utf8"`

> 	NodeName           string `asn1:"utf8"`

Reviewer 6: We included Node UID in https://github.com/kubernetes/kubernetes/pull/120780/files#diff-0c530c4dd8479db341cd002079be7ab62e34d49b42540785df9ad17431c8f108R243 and should probably do the same here.

Reviewer 2: Are all these fields required? If they are, I think we might one day regret not making the node name / uid optional.

Author: In the extension data itself, all fields are optional, in that you can always leave them as the empty string.  (I don't think the ASN.1 parser in Go tries to handle field presence or absence).

If the extension appears in a CertificateSigningRequest from a kubelet, the node restriction plugin will enforce that all the fields are populated, and consistent with reality.

If the extension appears in an issued certificate, I think it's up to the component that's validating the certificate to decide which fields it cares about.  For the specific example of certificates issued by the `kubernetes.io/kube-apiserver-client-pod` signer/approver, we probably want kube-apiserver to check all of the fields before letting the request past the authentication stage.

Reviewer 3: I think clearly documenting a minimum, non-negotiable subset of fields that are provably sufficient to guarantee unambiguous pod<->cert binding in Kubernetes is required, or including this ASN.1 OID field as an overall KEP requirement isn't much use.

Author: Sure, I can add a detailed proposal for exactly which fields need to be set in order for kube-apiserver to allow authentication via a certificate containing a PodIdentity extension.

Author: Based on discussion in API review, we are going to go with a dedicated PodCertificateRequest type, rather than encoding these details in the existing CertificateSigningRequest spec.request field.  This drastically lowers the importance of the PodIdentity extension, since it will only be used by the kube-apiserver-client-pod signer, not every third-party signer that wants to integrate with pod certificates.

}
```

> Certificates and CertificateRequests that embed this extension do not need to include any data in the standard X.509 Name or Subject Alternate Name fields.  The CertificateRequests generated by Kubelet will not, nor will the certificates issued by `kubernetes.io/kube-apiserver-client-pod`.

Reviewer 2: (nit / my opinion)
```suggestion
Certificates and certificate requests that embed this extension do not need to include any data in the standard X.509 name or subjectAlternateName fields.  The CertificateSigningRequests generated by the kubelet will not, nor will the certificates issued by `kubernetes.io/kube-apiserver-client-pod`.
```

Author: Done.

Utility functions for embedding and extracting this extension from X.509

> Certificate and CertificateRequest objects are defined in

Reviewer 2: If talking about objects from a particular Go library, name the library (eg `crypto/x509` from `std`). That makes the context clear to all readers.

---

Alternatively, if talking about data defined by X.509 itself:

```suggestion
Certificate and CertReqMessage objects are defined in
```
(I assume it's CertReqMessage and not CertReq, but haven't checked)

---

Author: I was talking about about crypto/x509.  Clarified.

`k8s.io/component-helpers/kubernetesx509`.  They are intended to be used both by
core Kubernetes components and third-party software that needs to interact with
these certificates.

### PodIdentity Extension Node Restriction

The noderestriction admission plugin will check the consistency of PodIdentity extensions in CertificateSigningRequest objects created by `system:node:` identities.  The creation request will only be admitted if the PodIdentity extension refers to a pod that is currently running on the node that requested the certificate.

This upholds the Kubernetes node isolation guarantee by ensuring that
compromising a single node only results in compromises of workloads that have
actually been scheduled onto that node.

<<[UNRESOLVED]>>

> Is it necessary or desirable to forbid non-Kubelet identities from creating CertificateSigningRequest objects that embed a PodIdentity extension?

Reviewer 6: Does anything prevent non-Kubelets from adding a Pod binding in TokenRequest, today? noderestriction doesn't appear to limit it, at least.

Author: I believe it is forbidden at some level.  But I guess we should just try it out.

<<[/UNRESOLVED]>>

> During Alpha, this capability is behind the PodIdentityX509Support feature gate.

Reviewer 2: ```suggestion
During Alpha, this capability would be behind the `PodIdentityX509Support` feature gate.
```

Author: Done.

### PodCertificate Projected Volume Sources

The PodCertificate projected volume source instructs Kubelet to inject a
credential bundle consisting of a private key and certificate chain into the
container fileystem.  Kubelet handles generating the key, requesting a
certificate from the named signer, and renewing the certificate before it
expires.

Kubelet supports only a few key generation strategies, each named by a string
that encapsulates both the key type and any parameters necessary (for example,
RSA modulus size).  The intention is to offer a a tasting menu of reasonable key
choices, rather than offering a flexibility to a wide variety of parameters.
Key types supported by Kubelet are "RSA2048", "RSA3072", "RSA4096", "ECDSAP256",
and "ECDSAP384".  If no key type is specified, Kubelet defaults to "ECDSAP256".

<<[UNRESOLVED]>>

> Should RSA2048 be offered?  I'm inclined to say no.  Are there any other key types that should be offered?

Reviewer 2: “I” isn't always clear - it's helpful to name who objects.

Author: Done.

<<[/UNRESOLVED]>>

Once the key is generated, Kubelet creates a CertificateSigningRequest that

> contains an X.509 CertificateRequest with no data in the Name or

Reviewer 2: ```suggestion
contains an X.509 CertReqMessage with no data in the Name or
```

> SubjectAlternateName fields, but instead embeds an X.509 PodIdentity extension

Reviewer 2: ```suggestion
subjectAlternateName fields, but instead embeds an X.509 Pod identity extension
```

Author: Done -- except PodIdentity is the name of proposed extension.

> with the details of the pod that mounted the volume.  The signer is intended to treat this CertificateRequest merely as an indication that it should issue a certificate for the named pod, rather than trying to issue a certificate that follows the structure of the CertificateRequest.

Reviewer 2: ```suggestion
with the details of the pod that mounted the volume. 
The signer is intended to treat this certificate request merely as an indication that it
should issue a certificate for the named pod, rather than trying to issue a certificate
that follows the structure of the CertificateRequest.
```
Also, can we explain why there's that mismatch between CSR and issued certificate?

Author: I'm a bit confused about what needs clarification; a Kubernetes signer implementation has never been required to pay any attention to the certificate request embedded in a CertificateSigningRequest object.  It's totally free to to issue whatever certificate it likes.

Reviewer 2: We haven't documented the [lack of] link between CertificateSigningRequest and certificates. https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#certificate-signing-requests implies that there is a link.

I know that the way these things work allow issuing an semi-related certificate that uses the same public key, but that's not the usual way people would interpret the name. So it's important to make it clear when “certificate signing request” actually means: “please sign some other certificate, not this one”.

> Kubelet then starts a single-object informer on the CertificateSigningRequest object it just created, and waits until the CSR is Approved and Issued.  It then takes the private key it generated and the issued certificate chain, and writes them to the pod filesystem in a single PEM file.  The first block in the PEM file contains the private key, and the remaining blocks contain the certificate chain in the same order they were in the CSR.

Reviewer 2: ```suggestion
The kubelet then starts a single-object informer on the CertificateSigningRequest object it just created, and waits until the CSR is Approved and Issued.  The kubelet then takes the private key it generated earlier, plus the issued certificate chain, and writes them to the filesystem of the relevant Pod volume, concatenated into a single PEM-encoded file.  The first block in the PEM file contains the private key, and the remaining blocks contain the certificate chain in the same order they were in the CSR.
```

Author: Done.

> Kubelet strips any inter-block comments and intra-block headers that were

Reviewer 2: ```suggestion
The kubelet strips any inter-block text comments and / or intra-block headers that were
```

Author: Done.

present in the issued certificate chain from the signer.

> Kubelet always sets a 24-hour TTL on its certificate requests, although of course the signer is free to ignore the requested TTL and return a certificate with any TTL.

Reviewer 2: ```suggestion
The kubelet always specifies a 24-hour TTL on its certificate requests, although of course the signer is free to ignore the requested TTL and return a certificate with any TTL.
```

Reviewer 2: Could specify that the kubelet accepts any certificate duration, or alternatively state that that kind of filtering is undefined.

:thought_balloon: I suppose someone could, eg, make an admission controller to reject Pod certificates that outlast the requested expiry by more than 24 hours.

Author: Clarified.  The draft impelemtnation enforces a minimum TTL, to prevent us from falling into an issuance loop.

During the initial volume mount, before the pod has started, any errors returned during the certificate issuance process will block pod startup.

> Kubelet will begin trying to rotate the certificate once less than 50% of its TTL is remaining, or after 24 hours, whichever is sooner.  Errors in the certificate issuance process are treated as non-blocking while the original certificate is still valid.  Once the original certificate has expired, errors are treated as blocking, because it's better for the pod to noisily go unhealthy than to continue to run with an expired certificate.

Reviewer 6: What does it mean that errors are "blocking," given that rotation is async? Does that mean we kill the Pod?

Reviewer 2: ```suggestion
The kubelet will begin trying to rotate the certificate once less than 50% of its TTL is remaining, or after 24 hours, whichever is sooner.  Errors in the certificate issuance process are treated as non-blocking while the original certificate is still valid.  Once the original certificate has expired, errors are treated as blocking, because it's better for the pod to noisily go unhealthy than to continue to run with an expired certificate.
```

Author: I just meant that we would start returning an error from the projected volume mount setup function, which is being polled every few seconds.  I don't know exactly what failure mode that will cause for the pod, but I believe it will be noisy at the very least (probably you'll get events emitted).

Author: Clarified to remove the "blocking/non-blocking" language.  Instead, I talk about whether or not errors are returned from the volume setup function.

<<[UNRESOLVED]>>
The rotation/renewal behavior is minimally tunable.  Is greater
tunability needed?  For example, is it necessary to let the pod author set the
TTL Kubelet passes in the CSR object, or to let the pod author control the
rotation safety threshold?
<<[/UNRESOLVED]>>

> 

Reviewer 2: How does the kubelet replace the PEM-encoded file / do we specify any level of atomicity?

Author: All projected volume content is written using an atomic symlink directory swap using the `atomicwriter` library.

Reviewer 2: Ok, cool.

+	// the request and issues a certificate chain, Kubelet writes the key and
+	// certificate chain to the pod filesystem.  The pod does not start until
+	// certificates have been issued for each podCertificate projected volume
+	// source in its spec.
+	//
+	// Kubelet will begin trying to rotate the certificate after 50% of the leaf
+	// certificate's lifetime has elapsed, or 24 hours, whichever is shorter.
+	//
+	// The credential bundle is a single file in PEM format.  The first PEM
+	// entry is the private key, and the remaining PEM entries are the
+	// certificate chain issued by the signer (typically, signers will return
+	// their certificate chain in leaf-to-root order).
+	//
+	// The named signer controls chooses the format of the certificate it
+	// issues; consult the signer implementation's documentation to learn how to
+	// use the certificates it issues.
+	//
+	// +featureGate=PodCertificateProjection
+ // +optional
+	PodCertificate *PodCertificateProjection `json:"podCertificate,omitempty" protobuf:"bytes,6,opt,name=podCertificate"`
}
```

#### Example certificate bundle written to the pod filesystem

```
-----BEGIN PRIVATE KEY-----

> <Omitted, so that automation stops mailing me>

Reviewer 2: ```suggestion
<Omitted, so that automation doesn't mail the KEP authors to complain>
```

Author: Done.

MRMwEQYDVQQDEwprdWJlcm5ldGVzMB4XDTIzMTAyOTA1NDU1NVoXDTIzMTAzMDA1
NTAyMlowADCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAMy7zINPgkDc
kCrb2EEuxvt6PLjthjnlq9uxl79dhQGl7EgIEgA4Kc+WKZ6uJsYJlJazhuc6egDY
7nr8s+MATtPV8hV6Rgd/8bMB2ktq1RVhEJ4kRAodSLDLnD4e0IhIZpXSK5DBCXF6
dRotiCEIGpCQmdMTDYZez6zTfN3E1yuhyw7bSKc2OJj6+MYly6oQwE0T7WVDNqoa
Ev2ZdHZjePkEhXvGRvm1fyxK4h0UJsaT38/COUW/Dal7Fx6VcJeY2QHI28W+NYP3
AOdx/7Hc6bbPcYZI/c1HPolUhoX6huiltGMn3yOLsaGjajwXUoE2aKsy3AHxeIkQ
4UlKQSabpb8CAwEAAaOBqDCBpTAOBgNVHQ8BAf8EBAMCA6gwEwYDVR0lBAwwCgYI
KwYBBQUHAwIwHwYDVR0jBBgwFoAUuQ7U2/jMkK4nY5R2rPvjyCiC2SQwXQYJKwYB
BAGDwlMBBFAwTgwHZGVmYXVsdAwHZGVmYXVsdAwHcGVtLXBvZAwkMjc3YjljNDkt
MzljMy00ZjIzLWIwYjEtZTg4OWVmOTBjMGIyDAtraW5kLXdvcmtlcjANBgkqhkiG
9w0BAQsFAAOCAQEAQd1G2CHdGPv9Lu9/SD7HPNXwI0K4JrXsWxJQ7Q16IVAuuWIY
mOdumO9L5PtTrV1widNm99oMEtztQHZoHHIkkypjfW+BB8XwO5er+FpcUCuvZ3w/
g9Zk9cZUqCdWER0UDl4RCLcS+M6QSq0nUg+JTZKRGJ5fjns3vZHA9Nswp4XRvjt8
yf6RY6oEJoa10QxB7YiNVmVkqLifhkQiF0Bmz7tONbsjceZ030Cs/wjdssnNzRGL
UdFfcDKw4QbMGqntJXPFqu86lk0IJwir+aTk/31yDZX1ZyOK0Y/SI7ZVVDeETTA9
k68gmm9HQCdRI3stW7TC3lB0Cd1XSSMUISIP0g==
-----END CERTIFICATE-----
```

### API Server PodIdentity Extension Support

> Kube-apiserver is extended to authenticate TLS client certificates that:

Reviewer 2: (nit)
```suggestion
The kube-apiserver is extended to authenticate TLS client certificates that:
```

Author: Done.

* Are signed by the existing `--client-ca-file` CA,
* Contain empty Name and SubjectAlternateName data, and
* Contain an X.509 PodIdentity extension.

These certificates are translated to an authenticator.Response that has:
* `User.Name` set to
  `system:serviceaccount:{PodIdentity.Namespace}:{PodIdentity.ServiceAccountName}`.
* `User.Extra["authentication.kubernetes.io/pod-name"]` set to `{PodIdentity.PodName}`

> * `User.Extra["authentication.kubernetes.io/pod-uid"]` set to `{PodIdentity.PodUID}`

Reviewer 6: Add node info here too.

Author: Done.

The net effect is that authenticating via a PodIdentity-bearing cert has the same object-bound behavior as authenticating via a bound service account token.

<<[UNRESOLVED]>>
Should we standardize an attribute like `authentication.kubernetes.io/node-name` and put the node name in there as well?
<<[/UNRESOLVED]>>

> During Alpha, this feature is gated by the PodIdentityX509Support feature flag.

Reviewer 2: ```suggestion
During Alpha, this feature is gated by the `PodIdentityX509Support` feature gate.
```
We call them feature gates. Too late to change now&hellip;

Author: Done.

### API Server Pod Client Certificate Approver / Signer

> Kube-controller-manager is extended to support a new signer name `kubernetes.io/kube-apiserver-client-pod`.

Reviewer 2: (nit)
```suggestion
The kube-controller-manager is extended to support a new certificate signer,
`kubernetes.io/kube-apiserver-client-pod`.
```

Author: Done.

The approver for this signer name automatically approves
CertificateSigningRequests that:
* Contain no Name or Subject Alternate Name data, 
* Embed a PodIdentity extension, and
* Are requested by node identities.

(The approver relies on the noderestriction plugin to verify that the pod and
node in the extension are consistent.)

The signer for this signer name issues certificates for all approved CSRs
addressed to the signer name.  The certificates it issues contain no Name or
Subject Alternate Name data, embed a PodIdentity extension, and have key usages
consistent with the certificate being used as a TLS client certificate. It uses the same CA key and certificates as `kubernetes.io/kube-apiserver-client`.

During Alpha, this feature is gated by the

> IssueKubeAPIServerPodClientCertificates feature flag.

Reviewer 2: ```suggestion
`IssueKubeAPIServerPodClientCertificates` feature gate.
```

Author: Done.

### Client-go InClusterConfig Enhancements

> Not prototyped yet --- TBD.

Reviewer 2: ```suggestion
<<[UNRESOLVED not prototyped yet ]>>
TBD
<<[/UNRESOLVED]>>
```

Author: Done.

burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [ ] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name:
  - Components depending on the feature gate:
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning

>     of a node?

Reviewer 2: I'd fill this in, even for `provisional`. Would save time tracking the gap later.

Author: Done.

    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.

> -->

Reviewer 2: Here are some ideas that I think we could sketch out:

- a **new subresource** for ServiceAccounts, that accepts a certificate signing request in some form, and (after checking) returns an issued certificate. The kubelet requests that subresource when creating a Pod that uses certificates to authenticate to the Kubernetes API
- [**HTTP Message Signatures**](https://datatracker.ietf.org/doc/draft-ietf-httpbis-message-signatures/) - still draft; good grounds to skip this alternative. But I would still list it for context.
- **HMAC-based one-time password** tokens for Pods; the token is never retransmitted, but the involved parties (Pod, kubelet, API server) have access to its cleartext, so it's not a big improvement over reused bearer tokens.

Why list alternatives? They help show that the approach we're proposing is sound and that we've really considered other ways to achieve the outcome.

Author: I've pulled this into an `unresolved` block in alternatives, and I will flesh them out.

* A PodCertificate projected volume source that instructs the kubelet to handle
  provisioning a key and certificate chain on behalf of a pod.
Taken together, this machinery makes it feasible to securely and automatically
deliver X.509 certificates to every pod in a cluster, without imposing an
unreasonable burden on application developers or cluster administrators.  

As a first application of this machinery, this KEP defines additional machinery that allows pods to automatically receive mTLS client certificates suitable for authenticating to kube-apiserver, with the same attributes provided by bound service account tokens:
* A PodIdentity X.509 extension to allow an X.509 certificate to carry all
  details of a pod's identity, including its UIDs and the node it is running on.
* A new TLS user-mapping capability in kube-apiserver to understand client
  certificates that embed PodIdentity extensions.
* A new PodCertificateRequest signer
  (`kubernetes.io/kube-apiserver-client-pod`), shipped in
  kube-controller-manager, to issue compatible client certificate to pods.
* An enhancement to client-go's InClusterConfig that automatically loads a

>   credential bundle from a well-known filesystem path and uses is as a client

Author: Done.

implementing and operating a third-party signer implementation today requires
intimate familiarity with Kubernetes' security model, otherwise your signer may
negate Kubernetes' node restriction security boundary.  This KEP moves
responsibility for enforcing node restriction on PodCertificateRequests into
kube-apiserver, so that signer implementations do not need to consider it.

### Non-Goals

**Specify a solution for pod-to-pod mTLS in core Kubernetes**: Pod-to-pod mTLS
(aka service mesh) is a rapidly-evolving area.  While all of the machinery
described in this KEP is expected (and is designed to be) useful for service
mesh implementations, it doesn't make sense to try and standardize pod-to-pod
mTLS at the core Kubernetes layer today.

**Support HSM-backed private keys**: In this proposal, private keys are
contained to the node (held only in kubelet memory and tmpfs projected volumes),
but they can still be extracted from a compromised workload (or compromised
node).  Storing the private keys in an HSM closes this loophole, but will require
an additional interface for kubelet to expose remote signing capabilities to workloads.

> **Be 100% compatible with software in the field today**: The primary example in this KEP is that the private key and certificate chain are delivered in a single credential-bundle file.  This is based on implementation and operation experience of a CSI-based certificates solution that delivered private keys and certificate chains in separate files, as expected by most commodity webservers today.  While this is an attractive idea, automatically rotating these separate files is fraught.  Even if both file's content is updated atomically, skewed reads in the workload can result in hard-to-diagnose errors if the workload reads the key before rotation and the certificate chain after rotation (or vice-versa).

Author: Done.

**Support identity delegation or impersonation**: We do not intend to support
issuing certificates that workloads can use as intermediate CA certificates in
order to issue further certificates, nor do we intend to support one privileged
workload minting certificates on behalf of other workloads.

## Design Details

### PodCertificateRequest Resource

PodCertificateRequest is a new API type in certificates.k8s.io/v1alpha1.  It's a stripped-down version of CertificateSigningRequest that is tailored directly to the pod certificate issuance use case.

PodCertificateRequest is fundamentally a statement that "Pod X is requesting a certificate from Signer Y".  The signer is in complete control of the form of the issued certificate.  As such, PodCertificateRequest objects don't contain a typical X.509 CSR.  Instead, they contain:
* A public key, PKIX-serialized.  This key must be an RSA key with modulus size
  3072 or 4096, or an ECDSA key with curve P256 or P384.
* A proof of possession of the corresponding private key.
  * If the key is an RSA key, then the signature is over the ASCII bytes of the

> 	  pod UID, using  RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented

Reviewer 7: ```suggestion
	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented
```

Author: Done.

type PodCertificateRequest struct {
	metav1.TypeMeta `json:",inline"`
	// +optional
	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`

	// spec contains the details about the certificate being requested.
	Spec PodCertificateRequestSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`

	// status contains the issued certificate, and a standard set of conditions.
	// +optional
	Status PodCertificateRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
}

// PodCertificateRequestSpec describes the certificate request.  All fields are
// immutable after creation.
type PodCertificateRequestSpec struct {
	// signerName indicates the request signer.
	SignerName string `json:"signerName" protobuf:"bytes,1,opt,name=signerName"`

	// podName is the name of the pod into which the certificate will be mounted.
	PodName string `json:"podName" protobuf:"bytes,2,opt,name=podName"`
	// podUID is the UID of the pod into which the certificate will be mounted.
	PodUID types.UID `json:"podUID" protobuf:"bytes,3,opt,name=podUID"`

	// serviceAccountName is the name of the service account the pod is running as.

> 	ServiceAccountName string `json:"serviceAccountName" protobuf:"bytes,4,opt,name=serviceAccountName"`

Author: Done.  It was feasible to populate the UIDs in the original PodCertificateRequest posted from Kubelet, so I've done that.  The node restriction admission plugin will validate them.

	// nodeName is the name of the node the pod is assigned to.
	NodeName string `json:"nodeName" protobuf:"bytes,6,opt,name=nodeName"`

	// pkixPublicKey is the PKIX-serialized public key the signer should issue
	// the certificate to.
	//

> 	// The key must be one of RSA-3072, RSA-4096, ECDSA-P256 or ECDSA-P384.

Reviewer 8: clarify this list of key types could be expanded in the future, and what a signer should do if it encountered a key type it does not support

Author: Done.

	PKIXPublicKey []byte `json:"pkixPublicKey" protobuf:"bytes,8,opt,name=pkixPublicKey"`

	// proofOfPossession proves that the requesting kubelet holds the private
	// key corresponding to pkixPublicKey.
	//
	// If the key is an RSA key, then the signature is over the ASCII bytes of
	// the pod UID, using  RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as
	// implemented by the golang function crypto/rsa.SignPKCS1v15).
	//
	// If the key is an ECDSA key, then the signature is as described by [SEC 1,
	// Version 2.0](https://www.secg.org/sec1-v2.pdf) (as implemented by the
	// golang library function crypto/ecdsa.SignASN1)

> 	ProofOfPossession []byte `json:"proofOfPossession" protobuf:"bytes,9,opt,name=proofOfPossession"`

Author: Done.

}

type PodCertificateRequestStatus struct {
	// conditions applied to the request. Known conditions are "Denied" and "Failed".
	// +listType=map
	// +listMapKey=type
	// +optional

> 	Conditions []PodCertificateRequestCondition `json:"conditions,omitempty" protobuf:"bytes,1,rep,name=conditions"`

Reviewer 8: consider using standard metav1.Conditions

Author: Done.

type PodCertificateRequestConditionType string

// Well-known condition types for certificate requests.
const (
	// Denied indicates the request was denied by the signer.
	PodCertificateRequestDenied PodCertificateRequestConditionType = "Denied"
	// Failed indicates the signer failed to issue the certificate.
	PodCertificateRequestFailed PodCertificateRequestConditionType = "Failed"
)

// PodCertificateRequestCondition describes a condition of a PodCertificateRequest object.
type PodCertificateRequestCondition struct {
	// type of the condition. Known conditions are "Denied" and "Failed".
	//
	// A "Denied" condition is added via the /status subresource, indicating the
	// request was denied.  Denied conditions cannot be removed once added.
	//
	// A "Failed" condition is added via the /status subresource, indicating the
	// signer failed to issue the certificate.  Failed conditions cannot be
	// removed once added.
	//
	// Only one condition of a given type is allowed.
	Type PodCertificateRequestConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RequestConditionType"`
	// status of the condition, one of True, False, Unknown.

> 	// Approved, Denied, and Failed conditions may not be "False" or "Unknown".

Author: Done.

	// +optional
	Message string `json:"message,omitempty" protobuf:"bytes,3,opt,name=message"`
	// lastUpdateTime is the time of the last update to this condition
	// +optional
	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty" protobuf:"bytes,4,opt,name=lastUpdateTime"`
	// lastTransitionTime is the time the condition last transitioned from one status to another.
	// If unset, when a new condition type is added or an existing condition's status is changed,
	// the server defaults this to the current time.
	// +optional
	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,5,opt,name=lastTransitionTime"`
}
```

### PodCertificate Projected Volume Sources

The PodCertificate projected volume source instructs kubelet to inject a
credential bundle consisting of a private key and certificate chain into the
container fileystem.  The kubelet handles generating the key, requesting a
certificate from the named signer, and renewing the certificate before it
expires.

The kubelet supports only a few key generation strategies, each named by a
string that encapsulates both the key type and any parameters necessary (for

> example, RSA modulus size).  The intention is to offer a a tasting menu of

Reviewer 7: ```suggestion
example, RSA modulus size).  The intention is to offer a tasting menu of
```

Author: Done.

   don't understand.

The available options were:

* Label CertificateSigningRequests with node/pod labels by convention.  This didn't meet goal 3.
* Hoist pod fields into CSR spec.  This didn't meet goal 4.
* Embed pod details into CSR spec.request field.  This didn't meet goals 1 and 2.
* New dedicated PodCertificateRequest type. Meets all goals.

#### API Definition

```go

// PodCertificateRequest encodes a pod requesting a certificate from a given
// signer.
//
// Kubelets use this API to implement podCertificate projected volumes
type PodCertificateRequest struct {
	metav1.TypeMeta `json:",inline"`
	// +optional
	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`

	// spec contains the details about the certificate being requested.
	Spec PodCertificateRequestSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`

	// status contains the issued certificate, and a standard set of conditions.
	// +optional
	Status PodCertificateRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
}

// PodCertificateRequestSpec describes the certificate request.  All fields are
// immutable after creation.

> type PodCertificateRequestSpec struct {

Reviewer 8: Consider whether a pod request for a certificate should be able to express a hint for certificate lifetime, similar to https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2784-csr-duration#motivation

A minimum (maybe the same as CSR of 10 minutes) to avoid the kubelet churning on expiring certificates could make sense.

	// a reason of UnsupportedKeyType.  It may also suggest a key type that it
	// does support by attaching an additional SuggestedKeyType condition, with
	// its reason field set to the suggested key type identifier.
	PKIXPublicKey []byte `json:"pkixPublicKey" protobuf:"bytes,8,opt,name=pkixPublicKey"`

	// proofOfPossession proves that the requesting Kubelet holds the private
	// key corresponding to pkixPublicKey.
	//
	// kube-apiserver validates the proof of possession during creation of the
	// PodCertificateRequest.
	//
	// If the key is an RSA key, then the signature is over the ASCII bytes of
	// the pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as
	// implemented by the golang function crypto/rsa.SignPKCS1v15).
	//
	// If the key is an ECDSA key, then the signature is as described by [SEC 1,
	// Version 2.0](https://www.secg.org/sec1-v2.pdf) (as implemented by the
	// golang library function crypto/ecdsa.SignASN1)
	ProofOfPossession []byte `json:"proofOfPossession" protobuf:"bytes,9,opt,name=proofOfPossession"`
}

type PodCertificateRequestStatus struct {
	// conditions applied to the request. Known conditions are "Denied",

> 	// "Failed", and "SuggestedKeyType".

Author: Done.

	// certificateChain is populated with an issued certificate by the signer.
	// This field is set via the /status subresource. Once populated, this field
	// is immutable.
	//
	// If the certificate signing request is denied, a condition of type
	// "Denied" is added and this field remains empty. If the signer cannot
	// issue the certificate, a condition of type "Failed" is added and this
	// field remains empty.
	//
	// Validation requirements:
	//  1. certificateChain must contain one or more entries.
	//  2. Each entry must contain a single DER-encoded ASN.1 Certificate structure as described in section 4 of RFC5280.
	//
	// If more than one entry is present, and the definition of the requested
	// spec.signerName does not indicate otherwise, the first entry is the
	// issued certificate, and subsequent entries should be treated as
	// intermediate certificates and presented in TLS handshakes.  When
	// projecting the chain into a pod volume, kubelet will preserve the order
	// of certificateChain.
	//
	// As a reminder, when serialized as JSON or YAML, each entry is
	// additionally base64-encoded.
	//
	// +listType=atomic
	// +optional

> 	CertificateChain [][]byte `json:"certificateChain,omitempty" protobuf:"bytes,2,opt,name=certificateChain"`

Reviewer 8: If the kubelet will need to be able to project intermediates distinct from the leaf cert, having `certificate []byte` and `intermediates [][]byte` might make more sense

If the kubelet will only inject PEM format, is there a benefit to making this different than the ultimately injected value? Everywhere else we have certificates in the REST API or config types we use PEM format (either directly as string or base64-encoded as []byte)

Author: There's no use case for projecting the cert separate from intermediates.  A properly-functioning client will always present them all, in the same order that they were projected into the filesystem.

As for PEM vs raw-DER, my thinking was that humans are never going to be requesting these certificates, so there's not much point in using PEM for the intermediate representation.  So we might as well save on some unneeded base64 round-tripping.

I don't fell strongly either way... this field could also be `CertificateChain string`, and just contain a bunch of PEM blocks.

Author: Done.

	// issuedAt is the time at which the signer issued the certificate.  This
	// field is set via the /status subresource.  Once populated, it is
	// immutable.
	//
	// +optional

> 	IssuedAt metav1.Time `json:"issuedAt,omitempty" protobuf:"bytes,3,opt,name=issuedAt"`

Author: Done.

	// notBefore is the time at which the certificate becomes valid.  This field
	// is set via the /status subresource.  Once populated, it is immutable.
	//
	// +optional
	NotBefore metav1.Time `json:"notBefore,omitempty" protobuf:"bytes,4,opt,name=notBefore"`

	// beginRefreshAt is the time at which the kubelet should begin trying to
	// refresh the certificate.  This field is set via the /status subresource,
	// and must be set at the same time as certificateChain.  Once populated,
	// this field is immutable.
	//
	// +optional

> 	BeginRefreshAt metav1.Time `json:"beginRefreshAt,omitempty" protobuf:"bytes,5,opt,name=beginRefreshAt"`

Author: Done.

- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
  - [ ] e2e Tests for all Beta API Operations (endpoints)
  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]

> - [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Author: Done (I think)

If the PodCertificateRequest is denied, the kubelet treats this as a fatal error and returns an error message from the volume setup.  It is assumed that if a signer marks a PodCertificateRequest as denied, then retrying the request will not make it change its mind.  The Pod will not start up.

If the PodCertificateRequest is failed, the kubelet also treats this as a
permanent error (although a Failed status does not necessarily indicate a
permanent problem with the chosen signer).  We assume that most pods are created
by a higher-level workload controller (Deployment, Job, DaemonSet), which will
supply the necessary retries by creating a new Pod.

The kubelet will begin trying to rotate the certificate at the time indicated by
the signer via the `PodCertificateRequest.Status.BeginRefreshAt` timestamp.
The kubelet will apply a randomized jitter of up to five minutes, to prevent
multiple volumes from gradually syncing their refresh periods. Errors in the
certificate issuance process are treated as non-fatal (logged but not returned
from the volume SetUp function) while the original certificate is still valid.

> Once the original certificate has expired, errors are will be returned from the

Reviewer 9: ```suggestion
Once the original certificate has expired, errors will be returned from the
```

Author: Done.

      signer->api: Update /status<br>Set CertificateChain, BeginRefreshAt
      kubelet->api: See PodCertificateRequest Y issued
      note over kubelet: Write credential bundle to pod filesystem
    end 
```

#### API Object Diff

```go
// PodCertificateProjection provides a private key and X.509 certificate in the
// pod filesystem.
type PodCertificateProjection struct {
	// Kubelet's generated CSRs will be addressed to this signer.
	SignerName string `json:"signerName,omitempty" protobuf:"bytes,1,rep,name=signerName"`

	// The type of keypair Kubelet will generate for the pod.
	//
	// Valid values are "RSA3072", "RSA4096", "ECDSAP256",
	// "ECDSAP384", and "ED25519".  If left empty, Kubelet defaults to "ECDSAP256".
	KeyType string `json:"keyType,omitempty" protobuf:"bytes,2,rep,name=keyType"`

	// Write the credential bundle at this path in the projected volume.
	//
	// The credential bundle is a single file that contains multiple PEM blocks.
	// The first PEM block is a PRIVATE KEY block, containing a PKCS#8 private
	// key.
	//
	// The remaining blocks are CERTIFICATE blocks, containing the issued
	// certificate chain from the signer (leaf and any intermediates).
	//

> 	// Prefer using credentialPath to keyPath and certificatePath, because the

Author: Done.

4UlKQSabpb8CAwEAAaOBqDCBpTAOBgNVHQ8BAf8EBAMCA6gwEwYDVR0lBAwwCgYI
KwYBBQUHAwIwHwYDVR0jBBgwFoAUuQ7U2/jMkK4nY5R2rPvjyCiC2SQwXQYJKwYB
BAGDwlMBBFAwTgwHZGVmYXVsdAwHZGVmYXVsdAwHcGVtLXBvZAwkMjc3YjljNDkt
MzljMy00ZjIzLWIwYjEtZTg4OWVmOTBjMGIyDAtraW5kLXdvcmtlcjANBgkqhkiG
9w0BAQsFAAOCAQEAQd1G2CHdGPv9Lu9/SD7HPNXwI0K4JrXsWxJQ7Q16IVAuuWIY
mOdumO9L5PtTrV1widNm99oMEtztQHZoHHIkkypjfW+BB8XwO5er+FpcUCuvZ3w/
g9Zk9cZUqCdWER0UDl4RCLcS+M6QSq0nUg+JTZKRGJ5fjns3vZHA9Nswp4XRvjt8
yf6RY6oEJoa10QxB7YiNVmVkqLifhkQiF0Bmz7tONbsjceZ030Cs/wjdssnNzRGL
UdFfcDKw4QbMGqntJXPFqu86lk0IJwir+aTk/31yDZX1ZyOK0Y/SI7ZVVDeETTA9
k68gmm9HQCdRI3stW7TC3lB0Cd1XSSMUISIP0g==
-----END CERTIFICATE-----
```

### Kubernetes API client pod certificates

#### X.509 PodIdentity Extension

The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
identity information into an X.509 certificate or certificate request, we can
define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
define a management policy for the PEN) and specify that it contains the
following DER-encoded structure:

```go

> type podIdentityASN1 struct {

Author: Added.  I don't actually know ASN.1, so it might be wrong.  I added some hedging language.

	Namespace          string `asn1:"utf8"`
	ServiceAccountName string `asn1:"utf8"`
  ServiceAccountUID  string `asn1:"utf8"`
	PodName            string `asn1:"utf8"`
	PodUID             string `asn1:"utf8"`
	NodeName           string `asn1:"utf8"`
  NodeUID            string `asn1:"utf8"`
}
```

Utility functions for embedding and extracting this extension from `crypto/x509`
Certificate objects are defined in `k8s.io/component-helpers/kubernetesx509`.

This extension, and the utility functions for manipulating it, are developed for
use by core Kubernetes components, although third parties could make use of

> them.  The extension should only be changed in backwards-compatible ways.

Author: Strictly speaking, we don't need to mark the extension as critical.  The only other content of the certificate will be some random bytes in the Subject.Name field

> them.  The extension should only be changed in backwards-compatible ways.

Author: To close on this, I think we *will* mark the extension as critical.  If we need to evolve the contents of the extension in the future, we will need to do so taking into consideration the behavior of TLS libraries when parsing the extension.

mechanism.

#### Client-go InClusterConfig Enhancements

If client-go `InClusterConfig()` detects a credential bundle file at `/var/run/secrets/kubernetes.io/serviceaccount/credentialbundle.pem`, it will configure itself to use the private key and client certificate chain from the bundle, periodically reloading them to catch when they are rotated.

### Risks and Mitigations

**Scalability**: In the current design and draft implementation, kubelet holds
the private keys for each pod on the node in memory.  If kubelet restarts, then
the private keys are lost, and kubelet will need to immediately generate new
keys and request new certificates for each volume on the pod.  If kubelet
crashloops, this can generate high CertificateSigningRequest traffic on
kube-apiserver.  This can potentially be mitigated by giving kubelet the
capability to read back the existing key from the projected volume on startup.

### User Stories

#### Story 1

I'm an application developer building a Kubernetes controller that I want to deploy into my cluster.  I want to make sure that my controller securely authenticates to kube-apiserver using mTLS.

> I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `noAutomountServiceAccountToken` because I want to ensure that my pod cannot use bearer tokens for authentication.

Reviewer 9: nit:
```suggestion
I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `automountServiceAccountToken=false` because I want to ensure that my pod cannot use bearer tokens for authentication.
```

Author: It's possible to use both, and I could imagine pods might want both a certificate (to talk to kube-apiserver), and a selection of service-account tokens (to talk to external APIs that only understand OIDC).

Author: Fixed the field name, thanks.

              fieldPath: metadata.namespace
```

I ensure that my pod uses client-go's InClusterConfig, and client-go automatically detects the presence uses the mTLS client certificate to authenticate to kube-apiserver.

#### Using this feature as a third-party signer implementer

### Future expansion: HSM support for private keys

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

> ##### Unit tests

Reviewer 9: Would be nice to at least try to list the planned tests below.

Author: Done.

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>

> ### Graduation Criteria

Reviewer 9: Clear graduation criteria are missing.

Author: I have added alpha graduation criteria.

GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

> ### Version Skew Strategy

Reviewer 9: Your feature touches apiserver, controller manager and kubelet, so it seems reasonable to try to outline the skew strategy.

Author: Added an explanation of the various cases of skew.  I need to confirm one detail, which I have marked with an unresolved block.

  CRI or CNI may require updating that component before the kubelet.
-->

## Production Readiness Review Questionnaire

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--

> This section must be completed when targeting alpha to a release.

Reviewer 9: This whole section is required for alpha, and you only provided answers for the first question.

Author: Oops, filled it out.

-->

###### How can this feature be enabled / disabled in a live cluster?

- Feature Gate: PodCertificateRequests
  - Components depending on the feature gate: kube-apiserver, kube-controller-manager
- Feature Gate: PodCertificateProjection

>   - Components depending on the feature gate: kubelet

Reviewer 9: In the document you've also mentioned the following two for kube-apiserver:
- `PodCertificateKubeApiserverClientAcceptance`
- `PodCertificateKubeApiserverClientIssuance`
but I don't seem them listed here nor in the `kep.yaml`.

Author: Rationalized this with the yaml.

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

###### What happens if we reenable the feature if it was previously rolled back?

###### Are there any tests for feature enablement/disablement?

<!--
The e2e framework does not currently support enabling or disabling feature
gates. However, unit tests in each component dealing with managing data, created
with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.

Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

### Rollout, Upgrade and Rollback Planning

<!--

> This section must be completed when targeting beta to a release.

Reviewer 9: Below sections although not required and won't block alpha are worthy of your time to try to tackle some of the answers. I believe you've already pointed out some issues with scalability after node restart.

Reviewer 9: This holds, especially the rollout when you have api enablement + 3 various feature gates. Although I believe the overall recommendation will be to turn all at once.

## Design Details

### PodCertificateRequest Resource

PodCertificateRequest is a new API type in certificates.k8s.io/v1alpha1.  It's a stripped-down version of CertificateSigningRequest that is tailored directly to the pod certificate issuance use case.

PodCertificateRequest is fundamentally a statement that "Pod X is requesting a certificate from Signer Y".  The signer is in complete control of the form of the issued certificate.  As such, PodCertificateRequest objects don't contain a typical X.509 CSR.  Instead, they contain:
* A public key, PKIX-serialized.  This key must be an RSA key with modulus size
  3072 or 4096, an ECDSA key with curve P256 or P384, or an ED25519 key.
* A proof of possession of the corresponding private key.
  * If the key is an RSA key, then the signature is over the ASCII bytes of the
	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented
	  by the golang function crypto/rsa.SignPKCS1v15).
  * If the key is an ECDSA key, then the signature is as described by [SEC 1,
	  Version 2.0](https://www.secg.org/sec1-v2.pdf) (as implemented by the golang
	  library function crypto/ecdsa.SignASN1)

>   * If the key is an ED25519 key, then the signature is as described b

Reviewer 10: What's "b" here? Is that referring the same as the ECDSA key?

Author: Sorry, that's a mistake.  I just fixed it.

PodCertificateRequest objects have a simplified lifecycle compared to
CertificateSigningRequests.  There are no separate approval and signing stages.
They can only be issued (and thus implicitly approved), denied (represented by a "Denied" status), or failed (represented by a "Failed" status).

Similar to CertificateSigningRequests, controllers that want to issue a
certificate in response to a PodCertificateRequest need to have two separate RBAC permissions:
* `update` on `podcertificaterequests/status`.
* `sign` on the requested signer.

Once a signer controller decides that a PodCertificateRequest is addressed to it by matching on the signer name, it should take one of the following actions:
* Issue the certificate by updating the status subresource of the
  PodCertificateRequest, setting both the `certificateChain` and
  `beginRefreshAt` fields.
* Reject the certificate by updating the status subresource of the
  PodCertificateRequest, adding a "Denied" condition.
* Indicate failure in issuing the certificate by updating the status subresource
  of the PodCertificateRequest, adding a "Failed" condition.

> PodCertificateRequest validation logic will:

Reviewer 3: Can we be very clear here about what the _kubelet_ does versus what the _signer_ does, and who owns which step.

I like the idea of a PodCertificateRequest generally, but I am a bit concerned with statements like

> * Confirm that the public key is one of the supported key types.
> * Confirm that the proof-of-possession is valid.
> * Confirm that the issued chain (if one is set) consists of valid certificates, that do indeed form a chain to each other.

that we are effectively creating a new CSR-style API that by-design only the `kubelet` can ever use, and granting the kubelet too much responsibility here.

As it sits, unless you are OK with the specifics of how the `kubelet` attests, validates, and mints certs, the API is sort of useless, and your options are

1. Ignore this  entire API
2. Fork the kubelet

Additionally, I am not sure how much work we are really saving for signers here, by creating a new API and putting the kubelet as the lowest-common-denomintor between the API and the signers.
 
I would rather do something than nothing here, but I wonder if we would we be better served by separating the PodCertificateRequest/Status API stuff from the actual kubelet implementation restrictions?

That could be as simple as kubelet feature-flagging that leaves the PodCertificateRequest API in place but disables the kubelet-specific bits, so the signer (or an intermediary controller) could handle these things using the same API.

Author: I'm not quite sure I understand this concern, but I might be missing something.

As it stands, the only kubelet-specific part of the API is that kubelet has support for projected volumes that result in the creation of PodCertificateRequest objects.  Additionally, we'll probably check that PodCertificateRequest objects are created by `system:node:*` users.

It's an explicit goal that third-party signers should be able to plug in to this system, without having to reimplement anything.  They just need to provide a controller that listens for PodCertificateRequests that are directed to their signer name.  There should be no need to replace any part of the kubelet or kube-apiserver support.

Reviewer 9: I believe the validation described here is strictly about the apiserver-side validation, not the kubelet. The important distinction is that this kind of validation will happen only during create/update actions.
Below, there's a description of noderestriction admission plugin, which does a consistency check combining the PCR with an actual pod.

Author: Yes, Kubelet does no validation.  It just creates the PodCertificateRequest objects.

* Confirm that the public key is one of the supported key types.
* Confirm that the proof-of-possession is valid.
* Confirm that the issued chain (if one is set) consists of valid certificates,
  that do indeed form a chain to each other.
* Confirm that neither or both of the certificateChain and beginRefreshAt
  timestamp are set.
* Ensure that all fields in PodCertificateRequest.Spec are immutable.
* Ensure that all fields in PodCertificateRequest.Status are immutable once the
  request has been issued, denied, or failed.

The noderestriction admission plugin checks the consistency of the information in created PodCertificateRequest objects:
* The namespace, pod name/UID, service account name/UID, and node name/UID must
  correspond to a real pod that currently exists in the cluster.
* The pod should be in the Pending or Running phases.
* The PodCertificateRequest must have been created by the user
  `system:node:foo`, where `foo` is the node named in the PodCertificateRequest
  spec.

This upholds the Kubernetes node isolation guarantee by ensuring that
compromising a single node only results in compromises of workloads that have
actually been scheduled onto that node.

> During alpha, the ability to store PodCertificate objects (and the node

Reviewer 3: This is maybe answering parts of my concern above, actually

Reviewer 9: An important note: feature gates are only for the development phase, once this feature reaches GA they will be removed.

restriction admission logic) will be gated by the PodCertificateRequests feature
gate.

#### Why a New Type?

The following constraints led to the adoption of a new API type, as opposed to extending CertificateSigningRequest:
1. We wanted the pod details to be legible without having to parse an ASN.1 CSR.
2. We wanted the pod details (especially the associated node name) to be

>   available for use in an informer selector.  While these could be surfaced as labels on CertificateSigningRequest objects, labels are immutable.

Reviewer 10: Do you mean the labels are mutable here?

Reviewer 9: This holds.

Author: Yes, I meant to say labels are *mutable*.  Updating.

Author: Done.

	// doesn't support the key type of publicKey.
	PodCertificateRequestConditionUnsupportedKeyType string = "UnsupportedKeyType"
)
```

### PodCertificate Projected Volume Sources

The PodCertificate projected volume source instructs kubelet to inject a
credential bundle consisting of a private key and certificate chain into the
container fileystem.  The kubelet handles generating the key, requesting a
certificate from the named signer, and renewing the certificate before it
expires.

The kubelet supports only a few key generation strategies, each named by a
string that encapsulates both the key type and any parameters necessary (for
example, RSA modulus size).  The intention is to offer a tasting menu of
reasonable key choices, rather than offering a flexibility to a wide variety of
parameters. Key types supported by kubelet are "RSA3072", "RSA4096",
"ECDSAP256", "ECDSAP384", and "ED25519".  If no key type is specified, kubelet
defaults to "ECDSAP256".

Once the key is generated, kubelet creates a PodCertificateRequest with the details of the pod that is mounting the volume.  The kubelet then waits for the PodCertificateRequest to be issued, denied, or failed.

> If the PodCertificateRequest is issued, then the kubelet writes both the generated private key and the issued certificate chain into the pod filesystem as a single file with multiple PEM blocks.  The first block in the PEM file contains the private key, and the remaining blocks contain the certificate chain in the same order they were in the PodCertificateRequest.

Reviewer 3: I am a little concerned with storing the privkey inline with the cert chain? I feel like that just promotes people sending around this single file with the cert chain and maybe not realizing it also contains the private key.

Might be better to separate the private key and chain into explicitly different files.

Author: This is an option in the configuration --- the pod author can choose a single file, or separate files for key and certificates.

  narrowly scoped to the job of issuing certificates directly to pods.  A
  PodCertificateRequest is fundamentally a statement that "pod X is requesting a
  certificate from signer Y".
* A PodCertificate projected volume source that instructs the kubelet to handle
  provisioning a key and certificate chain on behalf of a pod.
Taken together, this machinery makes it feasible to securely and automatically
deliver X.509 certificates to every pod in a cluster, without imposing an
unreasonable burden on application developers or cluster administrators.  

As a first application of this machinery, this KEP defines additional machinery
that allows pods to automatically receive mTLS client certificates suitable for
authenticating to kube-apiserver, with the same attributes provided by bound
service account tokens:
* A PodIdentity X.509 extension to allow an X.509 certificate to carry all
  details of a pod's identity, including its UIDs and the node it is running on.
* A new TLS user-mapping capability in kube-apiserver to understand client
  certificates that embed PodIdentity extensions.
* A new PodCertificateRequest signer

>   (`kubernetes.io/kube-apiserver-client-pod`), shipped in

Author: Sure, but I don't think "we will adequately document the feature" is something we need to state in the KEP.

>   (`kubernetes.io/kube-apiserver-client-pod`), shipped in

Reviewer 11: No I mean, actually document the signer here, in the KEP itself.

>   (`kubernetes.io/kube-apiserver-client-pod`), shipped in

Author: It is?  In the section "API Server Pod Client Certificate Approver / Signer".

tokens at any step**:  One of the most significant benefits of mTLS
authentication is that the credential is split into two pieces (private key and
certificate), only one of which is ever transmitted to the counterparty (and
thus subject to eavesdropping or theft by the counterparty).  In order to
preserve this guarantee, it's important that bearer tokens are not involved *at
any step*, otherwise the security of the scheme devolves to just being a bearer
token.  In this KEP, as long as both kubelet and the signer implementation
authenticate to kube-apiserver using their own separately-bootstrapped client
certificates, then no bearer tokens are involved in certificate issuance.

**Make it easy to ship secure third-party signer implementations**:  Securely
implementing and operating a third-party signer implementation today requires
intimate familiarity with Kubernetes' security model, otherwise your signer may
negate Kubernetes' node restriction security boundary.  This KEP moves
responsibility for enforcing node restriction on PodCertificateRequests into
kube-apiserver, so that signer implementations do not need to consider it.

> ### Non-Goals

Reviewer 12: i think it's worth mentioning explicitly as a non-goal that this proposal does not cover signing serving certificates for pods. that seems obvious and covered in the pod-to-pod mTLS point, but implicitly.

Author: It's a non-goal for the built-in signer to issue serving certs (although we will probably mark the certs to allow server cert usage; it doesn't hurt anything).

However, it's an explicit goal for third-party signers to be able to use the PodCertificate projected volume machinery to be able to issue serving certificates to pods.

Author: Clarified in the text.

Author: > I would not expect us to mark these certs as anything but client certs.

Agreed.  As discussed on slack, we will only set the client cert usages.  This doesn't preclude users from using them as server certs; since they will need to write custom TLS validation code, they can just ignore the lack of the "server" usage bits.

Subject Alternate Name entries.

**Support identity delegation or impersonation**: We do not intend to support
issuing certificates that workloads can use as intermediate CA certificates in
order to issue further certificates, nor do we intend to support one privileged
workload minting certificates on behalf of other workloads.

## Design Details

### PodCertificateRequest Resource

PodCertificateRequest is a new API type in certificates.k8s.io/v1alpha1.  It's a
stripped-down version of CertificateSigningRequest that is tailored directly to
the pod certificate issuance use case.

PodCertificateRequest is fundamentally a statement that "Pod X is requesting a
certificate from Signer Y".  The signer is in complete control of the form of
the issued certificate.  As such, PodCertificateRequest objects don't contain a
typical X.509 CSR.  Instead, they contain:
* A public key, PKIX-serialized.  This key must be an RSA key with modulus size
  3072 or 4096, an ECDSA key with curve P256 or P384, or an ED25519 key.
* A proof of possession of the corresponding private key.
  * If the key is an RSA key, then the signature is over the ASCII bytes of the

> 	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented

Author: I don't know how important this is --- the primary security guarantees here come from ensuring that only the correct kubelet can request a certificate for a given pod.  The proof of possession is just to prevent (implausible) accidents, as far as I can tell from https://www.rfc-editor.org/rfc/rfc4211#appendix-C

> 	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented

Author: And https://www.rfc-editor.org/rfc/rfc4211#section-4 only says that the proof of possession should sign over some representation of the identity being requested.  It doesn't include a timestamp.

> 	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented

Reviewer 11: Okay per my reading of that RFC, I think the pod UID is sufficient when combined with all the other node admission checks (i.e. we know that this exact pod is scheduled on this exact node with this exact SA in the same namespace).

If a different node tried to replay the PoP, it would fail because that pod isn't scheduled to it.

	  by the golang function crypto/rsa.SignPKCS1v15).
  * If the key is an ECDSA key, then the signature is as described by [SEC 1,
	  Version 2.0](https://www.secg.org/sec1-v2.pdf) (as implemented by the golang
	  library function crypto/ecdsa.SignASN1)
  * If the key is an ED25519 key, then the signature is as described by the
    [ED25519 Specification](https://ed25519.cr.yp.to/) (as implemented by
    the golang library crypto/ed25519.Sign).
* Details of the requesting Pod's identity:

>   * Namespace (indicated by the namespace the PodCertificateRequest is created in).

Author: Sure, added.

>   * Namespace (indicated by the namespace the PodCertificateRequest is created in).

Author: Hmmm, actually, I'm not quite sure here.  The namespace is indicated by which namespace the PodCertificateRequest (and pod, and service account) are actually in

I'm not sure it makes sense to embed the UID of the namespace in the PodCertificateRequest, although we *could* have `kubernetes.io/kube-apiserver-client-pod`  embed the UID in the issued certificate.

  * Pod name
  * Pod UID
  * Service account name
  * Service account UID
  * Node name
  * Node UID
* The signer name from which a certificate is being requested.
* (In the status subobject): An issued certificate chain, stored as a single
  PEM-formatted string.  If there is more than one issued certificate, the first
  element of the chain is assumed to be the leaf certificate, followed by any
  necessary intermediates in leaf-to-root order.

> * (In the status subobject): A timestamp at which time the requester that made

Author: You mention this at two points, let's discuss it overall here.

While in my perfect world, all certificates would be short-lived, I don't think we can be this prescriptive for third-party signers.

For Kubernetes built-in signers, we control the code, so we can make sure the certificates are short-lived.  Additional validation isn't needed for our use cases.

If a third party wants to come and build a PodCertificateRequest signer called `my-cool-signer.com/lets-encrypt` that issues 90-day certs from LetsEncrypt, I don't think we should stop them.  They can assess the revocation story for their certs (which WebPKI certs will have) and decide whether or not its a good idea.

That's not to say we shouldn't do something about revocation within the framework of the Kubernetes API.  But it's a complicated problem with many sub-optimal solutions, and I don't think we have the time to figure out a workable approach before freeze.

For right now, a third-party signer implementation can bring their own revocation story.  It could be as simple as having their signing controller export the CRL in a configmap, and then making sure their workloads mount the configmap at a well-known pod.  To use their third-party certs, they'll have to be using their own third-party library code (not client-go), so they can make sure they plumb the CRL.

Secondly, even if we do decide to limit the lifetime of certs that can be issued through PodCertificateRequests, it should probably be part of validation on kube-apiserver, rather than in kubelet.

> * (In the status subobject): A timestamp at which time the requester that made

Reviewer 11: We discussed most of this offline but re: future revocation support - I think we could provide full support in client-go for an embedded CRL distributed via a mounted CTB.  I am not worried about it for alpha, but I do want to push towards a future where revocation is the norm and not the exception in our ecosystem.

> * (In the status subobject): A timestamp at which time the requester that made

Author: Agreed.  I've also updated the text to reflect our conversation.

Specifically:
* PodCertificateRequests and volume specs now carry `maxExpirationSeconds`.
* Kube-apiserver will restrict `maxExpirationSeconds` based on a per-signer configuration.
* Kube-apiserver will enforce that the signer actually issued a certificate that conforms to `maxExpirationSeconds`

  the PodCertificateRequest should begin trying to refresh the certificate.

PodCertificateRequest objects have a simplified lifecycle compared to
CertificateSigningRequests.  There are no separate approval and signing stages.
They can only be issued (and thus implicitly approved), denied (represented by a
"Denied" status), or failed (represented by a "Failed" status).

Similar to CertificateSigningRequests, controllers that want to issue a
certificate in response to a PodCertificateRequest need to have two separate
RBAC permissions:
* `update` on `podcertificaterequests/status`.

> * `sign` on the requested signer.

Author: That's my intention --- I will update the text.

PodCertificateRequest validation logic will:
* Confirm that the public key is one of the supported key types.
* Confirm that the proof-of-possession is valid.
* Confirm that the issued chain (if one is set) consists of valid certificates,
  that do indeed form a chain to each other.
* Confirm that neither or both of the certificateChain and beginRefreshAt
  timestamp are set.
* Ensure that all fields in PodCertificateRequest.Spec are immutable.
* Ensure that all fields in PodCertificateRequest.Status are immutable once the
  request has been issued, denied, or failed.

The noderestriction admission plugin checks the consistency of the information
in created PodCertificateRequest objects:
* The namespace, pod name/UID, service account name/UID, and node name/UID must
  correspond to a real pod that currently exists in the cluster.
* The pod should be in the Pending or Running phases.
* The PodCertificateRequest must have been created by the user
  `system:node:foo`, where `foo` is the node named in the PodCertificateRequest
  spec.

This upholds the Kubernetes node isolation guarantee by ensuring that
compromising a single node only results in compromises of workloads that have

> actually been scheduled onto that node.

Author: Oh, yeah, good point.  Fixed.

-->

Unit test list (informed by draft implementation):
* pkg/api/pod/util_test.go
* pkg/apis/certificates/validation_test.go
* pkg/apis/core/validation/validation_test.go
* pkg/controller/certificates/authority/authority_test.go
* pkg/controller/certificates/podcertificaterequest_controller_test.go
* pkg/controller/certificates/signer/signer_test.go
* pkg/kubelet/podcertificate/podcertificate_manager_test.go
* pkg/registry/certificates/podcertificaterequest/storage/storage_test.go
* pkg/registry/certificates/podcertificaterequest/strategy_test.go
* pkg/volume/projected/projected_test.go
* plugin/pkg/admission/noderestriction/admission_test.go
* plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy_test.go
* plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy_test.go
* staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go
* staging/src/k8s.io/client-go/rest/config_test.go
* staging/src/k8s.io/component-helpers/kubernetesx509/kubernetesx509_test.go

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those

> in the form of:

Reviewer 9: Nit: but the above comment explicitly requires both the directory and coverage :sweat_smile:

Author: The coverage tab crashes in Chrome :/ I need to dig up a Firefox install

Reviewer 9: You can always just run `go test -race -cover <package>` :wink:

Author: Will do.

-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

* test/e2e/auth/projected_podcertificate.go: Checks end-to-end issuance and
  usage of pod certificates issued by `kubernetes.io/kube-apiserver-client-pod`.
  The test will create an agnhost pod with a mounted credential bundle, and
  agnhost will use the credential bundle to connect to kube-apiserver.

### Graduation Criteria

For Alpha:
* The feature will be gated behind feature flags and alpha API
enablement.

> * An e2e test will exercise the feature in order to prevent

Reviewer 9: We don't enable alpha feature gates in our regular e2es, so make sure to add them to appropriate test suite which does it.

Author: Will do.  We had to do the same thing for ClusterTrustBundles.

  specify the new key type, but these pods will fail to start when actually
  scheduled onto a node.  Kubelet will continually return an error during volume
  set up, when it validates the specified key type against its list of supported
  key types.
* (`kube-apiserver>=N`, `kubelet<N`, `kube-controller-manager>=N`): Same as the
  previous case.
* (`kube-apiserver>=N`, `kubelet>=N`, `kube-controller-manager<N`):
  kube-apiserver will admit pods with PodCertificate projected volumes that
  specify the new key type.  These pods will still fail to start when scheduled
  onto a node.  Kubelet will successfully create PodCertificateRequest objects,
  but kube-controller-manager will continually reject them because it does not
  understand the new key type.  Kubelet will reflect this by returning a
  permanent error from volume set up.
* (`kube-apiserver>=N`, `kubelet>=N`, `kube-controller-manager>=N`): No
  problems, everything works.

Based on the above analysis, we may want to consider making kube-scheduler aware
of which kubelet versions support PodCertificate projected volumes (and which
key types are supported by version), and try to steer pods towards nodes that

> can successfully mount them.

Reviewer 9: We don't do that kind of checks anywhere I'm aware of, so I'd advise against it.

Reviewer 9: Rather make sure to provide sufficient information in the logs or through events which would allow a user to identify such a situation.

Author: Removed that paragraph.

token.  In this KEP, as long as both kubelet and the signer implementation
authenticate to kube-apiserver using their own separately-bootstrapped client
certificates, then no bearer tokens are involved in certificate issuance.

**Make it easy to ship secure third-party signer implementations**:  Securely
implementing and operating a third-party signer implementation today requires
intimate familiarity with Kubernetes' security model, otherwise your signer may
negate Kubernetes' node restriction security boundary.  This KEP moves
responsibility for enforcing node restriction on PodCertificateRequests into
kube-apiserver, so that signer implementations do not need to consider it.

**Support issuance of both client and server certificates**: While this KEP does
not include a concrete signer that issues dedicated server certificates,
third-party signer implementations should be able to do so on top of the
machinery described in this KEP.  Additionally, the certificates isssed by
`kubernetes.io/kube-apiserver-client-pod` specify key and certificate usages
that permit the certificates to be used for TLS servers, although they will not

> work with standard TLS server certificate validation.

Author: I don't feel strongly about this, but I included this for symmetry with tokens (which can be used for server identification).

My half-baked thought about why we should set the bits to allow server certificate usage:
* We have seen from Kubernetes service account JWTs that a rich ecosystem of usages for these JWTs (including both client and server authentication usages) has sprung up.
* Unlike tokens (which require a distinct audience for distinct purposes), certificates have no audience constraint (because they aren't bearer credentials, so they aren't vulnerable to the attacks that audience closes off).
* So the ecosystem that springs up around these certificates would likely collapse on using the default certificate (maybe by federating that CA that issued them into external identity systems).
* At least some of the users in that ecosystems would want to use the certificate as a server certificate.

> work with standard TLS server certificate validation.

Author: As documented above, we will omit the "server" usage.  Those who want to use the cert as a server cert will need to be writing custom TLS validation logic, and thus are free to ignore the lack of server usage bits.

	// +optional
	CertificateChain string `json:"certificateChain,omitempty" protobuf:"bytes,2,opt,name=certificateChain"`

	// issuedAt is the time at which the signer issued the certificate.  This
	// field is set via the /status subresource.  Once populated, it is
	// immutable.  The signer must set this field at the same time it sets 
  // certificateChain.
	//
	// +optional
	IssuedAt *metav1.Time `json:"issuedAt,omitempty" protobuf:"bytes,3,opt,name=issuedAt"`

	// notBefore is the time at which the certificate becomes valid.  This field
	// is set via the /status subresource.  Once populated, it is immutable.  The 
  // signer must set this field at the same time it sets certificateChain.
	//
	// +optional
	NotBefore *metav1.Time `json:"notBefore,omitempty" protobuf:"bytes,4,opt,name=notBefore"`

	// beginRefreshAt is the time at which the kubelet should begin trying to
	// refresh the certificate.  This field is set via the /status subresource,
	// and must be set at the same time as certificateChain.  Once populated,
	// this field is immutable.
  //
  // This field is only a hint.  Kubelet may start refreshing before or after
  // this time if necessary.
	//
	// +optional

> 	BeginRefreshAt *metav1.Time `json:"beginRefreshAt,omitempty" protobuf:"bytes,5,opt,name=beginRefreshAt"`

Author: This is explicitly a hint.  We address clock skew by having the signer backdate the cert for 5 minutes when it issues it.

> 	BeginRefreshAt *metav1.Time `json:"beginRefreshAt,omitempty" protobuf:"bytes,5,opt,name=beginRefreshAt"`

Author: Added.

	// The first PEM block is a PRIVATE KEY block, containing a PKCS#8 private
	// key.
	//
	// The remaining blocks are CERTIFICATE blocks, containing the issued
	// certificate chain from the signer (leaf and any intermediates).
	//
	// Using credentialBundlePath lets your Pod's application code make a single 
	// atomic read that retrieves a consistent key and certificate chain.  If you 
	// project them to separate files, your application code will need to 
	// additionally check that the leaf certificate was issued to the key.
	//
	// Mutually exclusive with keyPath and certificateChainPath.
	CredentialBundlePath string `json:"credentialBundlePath,omitempty" protobuf:"bytes,3,rep,name=credentialBundlePath"`

	// Write the key at this path in the projected volume.
	//
	// Mutually exclusive with credentialPath.
	//
	// When using keyPath and certificateChainPath, your application needs to check 
	// that the key and leaf certificate are consistent, because it is possible to 
	// read the files mid-rotation.
	KeyPath string `json:"keyPath,omitempty" protobuf:"bytes,4,rep,name=keyPath"`

	// Write the certificate chain at this path in the projected volume.
	//

> 	// Mutually exclusive with credentialPath.

Author: Fixed.

4UlKQSabpb8CAwEAAaOBqDCBpTAOBgNVHQ8BAf8EBAMCA6gwEwYDVR0lBAwwCgYI
KwYBBQUHAwIwHwYDVR0jBBgwFoAUuQ7U2/jMkK4nY5R2rPvjyCiC2SQwXQYJKwYB
BAGDwlMBBFAwTgwHZGVmYXVsdAwHZGVmYXVsdAwHcGVtLXBvZAwkMjc3YjljNDkt
MzljMy00ZjIzLWIwYjEtZTg4OWVmOTBjMGIyDAtraW5kLXdvcmtlcjANBgkqhkiG
9w0BAQsFAAOCAQEAQd1G2CHdGPv9Lu9/SD7HPNXwI0K4JrXsWxJQ7Q16IVAuuWIY
mOdumO9L5PtTrV1widNm99oMEtztQHZoHHIkkypjfW+BB8XwO5er+FpcUCuvZ3w/
g9Zk9cZUqCdWER0UDl4RCLcS+M6QSq0nUg+JTZKRGJ5fjns3vZHA9Nswp4XRvjt8
yf6RY6oEJoa10QxB7YiNVmVkqLifhkQiF0Bmz7tONbsjceZ030Cs/wjdssnNzRGL
UdFfcDKw4QbMGqntJXPFqu86lk0IJwir+aTk/31yDZX1ZyOK0Y/SI7ZVVDeETTA9
k68gmm9HQCdRI3stW7TC3lB0Cd1XSSMUISIP0g==
-----END CERTIFICATE-----
```

### Kubernetes API client pod certificates

#### X.509 PodIdentity Extension

The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
identity information into an X.509 certificate or certificate request, we can
define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
define a management policy for the PEN) and specify that it contains the
following DER-encoded structure:

```go
type podIdentityASN1 struct {

> 	Namespace          string `asn1:"utf8,tag:0"`

Author: Added

	ServiceAccountName string `asn1:"utf8,tag:1"`
	ServiceAccountUID  string `asn1:"utf8,tag:2"`
	PodName            string `asn1:"utf8,tag:3"`
	PodUID             string `asn1:"utf8,tag:4"`
	NodeName           string `asn1:"utf8,tag:5"`
	NodeUID            string `asn1:"utf8,tag:6"`
}
```

This extension is defined by the behavior of the go `encoding/asn1` package in
go 1.23.  A translation to ASN.1 notation is provided for convenience, but in
the case of conflict, the go version controls:
```asn1
PodIdentity SEQUENCE {
  namespace [0] UTF8String,
  serviceAccountName [1] UTF8String,
  serviceAccountUID [2] UTF8String,
  podName [3] UTF8String,
  podUID [4] UTF8String,
  nodeName [5] UTF8String,
  nodeUID [6] UTF8String
}
```

Utility functions for embedding and extracting this extension from `crypto/x509`

> Certificate objects are defined in `k8s.io/component-helpers/kubernetesx509`.

Author: Client-go doesn't need to ever decode these certs.  Only kube-apiserver needs to decode them, and only kube-controller-manager needs to encode them.  So component-helpers seemed like the correct home.

I don't have strong preferences on where this functionality lives, though.

> Certificate objects are defined in `k8s.io/component-helpers/kubernetesx509`.

Reviewer 11: Maybe @deads2k has some thoughts here?

This extension, and the utility functions for manipulating it, are developed for
use by core Kubernetes components, although third parties could make use of
them.  The extension should only be changed in backwards-compatible ways.

#### API Server PodIdentity Extension Support

The kube-apiserver is extended to authenticate TLS client certificates that
represent service account identities in the following format:

> * Are signed by root certificates provided by a new `--pod-client-ca-file` flag,

Author: Done.

* Contain no SubjectAlternateName data, and
* Contain an X.509 PodIdentity extension.

> These certificates are translated to an authenticator.Response that has:

Author: Added.

* `User.Name` set to
  `system:serviceaccount:{PodIdentity.Namespace}:{PodIdentity.ServiceAccountName}`.
* `User.UID` set to `PodIdentity.ServiceAccountUID`.
* `User.Extra["authentication.kubernetes.io/pod-name"]` set to
  `{PodIdentity.PodName}`
* `User.Extra["authentication.kubernetes.io/pod-uid"]` set to
  `{PodIdentity.PodUID}`
* `User.Extra["authentication.kubernetes.io/node-name"]` set to
  `{PodIdentity.NodeName}`
  * `User.Extra["authentication.kubernetes.io/node-uid"]` set to
  `{PodIdentity.NodeUID}`

The net effect is that authenticating via a PodIdentity-bearing cert has the
same object-bound behavior as authenticating via a bound service account token.

During Alpha, this feature will be gated by the
`PodCertificateKubeApiserverClientAcceptance` feature gate.

#### API Server Pod Client Certificate Approver / Signer

The kube-controller-manager is extended to support a new certificate signer,

> `kubernetes.io/kube-apiserver-client-pod`.

Reviewer 11: Or maybe we don't have to because these certs have no meaning outside of KAS <-> pod?

> `kubernetes.io/kube-apiserver-client-pod`.

Author: I don't think we need anything but reading from a file.  If a provider wants to store the root key somewhere else, they can disable this controller in kube-controller-manager and provider a reimplementation in another controller.

It's the same story as for the existing built-in signer.

The controller for this signer name automatically issues certificates for
PodCertificateRequests that are addressed to it.  It performs no additional
validation on PodCertificateRequests beyond the validation enforced by
kube-apiserver.

The certificates it issues are meant to be understood by kube-apiserver as
described above.  They:

> * Have a CommonName set to strong random text.  (This maintains flexibility, in

Author: Removed it.

  case we wish to include a meaningful common name in the future.)

> * Have a PodIdentity extension (non-critical) that contains the information from

Author: Done.

  `PodCertificateRequest.Spec`.
* Key usages consistent with the certificate being used as a TLS client

>   certificate or server certificate.

Author: Removed.

* It uses a CA specified by a new pair of private key / root certificate flags
  (although implementations are free to set this flag to point at the CA as
  another signer).
* It issues certificates that are valid for 24 hours, and tells Kubelet to begin
  refreshing them at 12 hours.

During Alpha, this feature is gated by the
`PodCertificateKubeApiserverClientIssuance` feature gate.

Because kube-apiserver will interpret the PodIdentity extension as expressing an
object binding between the presented certificate and the pod named in the
extension, once a pod is no longer running, all certificates that were issued to
it will become invalid.  This, along with the short lifetime of the
certificates, removes the need for a more permanent certificate revocation

> mechanism.

Reviewer 11: (actually probably makes more sense for KAS to enforce TTL via validation)

> mechanism.

Author: This mechanism has been added to the KEP.

#### Client-go InClusterConfig Enhancements

If client-go `InClusterConfig()` detects a credential bundle file at
`/var/run/secrets/kubernetes.io/serviceaccount/credentialbundle.pem`, it will
configure itself to use the private key and client certificate chain from the

> bundle, periodically reloading them to catch when they are rotated.

Author: Hmmm, maybe for alpha we should make it be a separate top-level function so that you have to explicitly opt-in?  Like `rest.InClusterPodCertificateConfig()`?

> bundle, periodically reloading them to catch when they are rotated.

Reviewer 11: I am not sure of the right approach, but I do want to give folks the ability to not leak go routines.

> bundle, periodically reloading them to catch when they are rotated.

Author: I've updated the kep with this approach.  In the future we can tackle trying to use these credentials by default.

### Risks and Mitigations

**Scalability**: In the current design and draft implementation, kubelet holds
the private keys for each pod on the node in memory.  If kubelet restarts, then
the private keys are lost, and kubelet will need to immediately generate new
keys and request new certificates for each volume on the pod.  If kubelet
crashloops, this can generate high CertificateSigningRequest traffic on
kube-apiserver.  This can potentially be mitigated by giving kubelet the
capability to read back the existing key from the projected volume on startup.

### User Stories

#### Story 1

I'm an application developer building a Kubernetes controller that I want to
deploy into my cluster.  I want to make sure that my controller securely
authenticates to kube-apiserver using mTLS.

I add a new projected volume to my pod, containing the information that

> client-go expects at a well-known path.  I set

Reviewer 11: It seems really awkward to require this very manual process, especially with magic well known paths on disk in play.

Author: In the future, we can think about automatically injecting these as a replacement for the existing kube-apiserver-access volumes.  I didn't want to cover that in Alpha because there's already enough work just landing the machinery without thinking about driving adoption.

However, I think there's a larger conversation to be had on this topic in SIG Auth.  My personal thoughts are:
* Automatically granting API access credentials by default is probably a mistake.  Most workloads in a Kubernetes cluster do not actually need to talk to the Kubernetes API.
* Mutating admission controllers (especially webhooks) cause a lot of problems.  In general, workload authors expect the spec they write to be the spec that gets loaded into the cluster.
* In a perfect world, I would make workload authors who want Kubernetes API access opt in by inserting a token (or pod certificate volume).

Reviewer 11: Yeah I am fine with this as-is, but lets discuss further in a future meeting.

`automountServiceAccountToken=false` because I want to ensure that my pod is not
using bearer tokens for authentication.
```yaml
apiVersion: v1
kind: Pod
metadata:
  namespace: default
  name: pod-certificates-example
spec:
  restartPolicy: OnFailure
  automountServiceAccountToken: false
  containers:
  - name: main
    image: debian
    command: ['sleep', 'infinity']
    volumeMounts:
    - name: kube-apiserver-client-certificate

>       mountPath: /var/run/secrets/kubernetes.io/api-mtls-access

Author: Sure.

  volumes:
  - name: kube-apiserver-client-certificate
    projected:
      sources:
      - podCertificate:
          signerName: "kubernetes.io/kube-apiserver-client-pod"

>           credentialBundlePath: client-credentials.pem

Author: Done.

* pkg/kubelet/podcertificate/podcertificate_manager_test.go
* pkg/registry/certificates/podcertificaterequest/storage/storage_test.go
* pkg/registry/certificates/podcertificaterequest/strategy_test.go
* pkg/volume/projected/projected_test.go
* plugin/pkg/admission/noderestriction/admission_test.go
* plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy_test.go
* plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy_test.go
* staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go
* staging/src/k8s.io/client-go/rest/config_test.go
* staging/src/k8s.io/component-helpers/kubernetesx509/kubernetesx509_test.go

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `<package>`: `<date>` - `<test coverage>`

> ##### Integration tests

Reviewer 11: I suspect we could manually run enough of the kcm/kubelet pieces during integration tests to exercise the various node restriction and new signer bits.

Once a signer controller decides that a PodCertificateRequest is addressed to it
by matching on the signer name, it should take one of the following actions:
* Issue the certificate by updating the status subresource of the
  PodCertificateRequest, setting both the `certificateChain` and
  `beginRefreshAt` fields.
* Reject the certificate by updating the status subresource of the
  PodCertificateRequest, adding a "Denied" condition.
* Indicate failure in issuing the certificate by updating the status subresource
  of the PodCertificateRequest, adding a "Failed" condition.

PodCertificateRequest validation logic will:
* Confirm that the public key is one of the supported key types.
* Confirm that the proof-of-possession is valid.
* Confirm that the issued chain (if one is set) consists of valid certificates,
  that do indeed form a chain to each other.
* Confirm that neither or both of the certificateChain and beginRefreshAt
  timestamp are set.
* Ensure that all fields in PodCertificateRequest.Spec are immutable.
* Ensure that all fields in PodCertificateRequest.Status are immutable once the
  request has been issued, denied, or failed.

> The `maxExpirationSeconds` field will have the following logic:

Reviewer 11: Probably need a min for this field -> does not make sense to let clients ask for 1 second certs.

Author: Good point.  Updated with a min of 1 hour.  It should be safe to drop that down in the future.

* kube-apiserver will be configurable with an optional maximum duration to allow
  for certificates issued via PodCertificateRequests, specified in the
  `--pod-certificate-request-max-duration` flag.  The syntax of this flag will
  support setting different maximum durations for different signers, with some
  level of wildcarding.  If a signer is not covered by the flag value,
  kube-apiserver will assume a maximum duration of 24 hours.
* If the creator of the PodCertificateRequest left `maxExpirationSeconds` at 0,

>   then kube-apiserver will default it to 24 hours.

Reviewer 11: follow-up -> decide if we need to support both a max exp seconds per signer and a default per signer.

  `X509SHA256=<sha256-hash-of-the-certificate>`.

The net effect is that authenticating via a PodIdentity-bearing cert has the
same object-bound behavior as authenticating via a bound service account token.

During Alpha, this feature will be gated by the
`PodCertificateKubeApiserverClientAcceptance` feature gate.

#### API Server Pod Client Certificate Signer

The kube-controller-manager is extended to support a new certificate signer,
`kubernetes.io/kube-apiserver-client-pod`.

The controller for this signer name automatically issues certificates for
PodCertificateRequests that are addressed to it.  It performs no additional
validation on PodCertificateRequests beyond the validation enforced by
kube-apiserver.

The certificates it issues are meant to be understood by kube-apiserver as
described above.  They:
* Have a PodIdentity extension (critical) that contains the information from
  `PodCertificateRequest.Spec`.
* Key usages consistent with the certificate being used as a TLS client
  certificate.
* It uses a CA specified by a new pair of private key / root certificate flags
  (although implementations are free to set this flag to point at the CA as
  another signer).

> * It issues certificates that are valid for 24 hours, and tells Kubelet to begin

Reviewer 11: What if a client asks for a shorter TTL?  What is the min TTL this signer supports?

Author: Good point.  I will update so that it issues certs for `min(maxExpirationSeconds, 24hours)`.
